// @credo-ts/didcomm/build/modules/routing/handlers/DidCommMediationRequestHandler.mjs
import { DidCommOutboundMessageContext } from "../../../models/DidCommOutboundMessageContext.mjs";
import "../../../models/index.mjs";
import { DidCommMediationRequestMessage } from "../messages/DidCommMediationRequestMessage.mjs";

//#region src/modules/routing/handlers/DidCommMediationRequestHandler.ts
var DidCommMediationRequestHandler = class {
	constructor(mediatorService, mediatorModuleConfig) {
		this.supportedMessages = [DidCommMediationRequestMessage];
		this.mediatorService = mediatorService;
		this.mediatorModuleConfig = mediatorModuleConfig;
	}
	async handle(messageContext) {
		const connection = messageContext.assertReadyConnection();
		const mediationRecord = await this.mediatorService.processMediationRequest(messageContext);
		if (this.mediatorModuleConfig.autoAcceptMediationRequests) {
			const { message } = await this.mediatorService.createGrantMediationMessage(messageContext.agentContext, mediationRecord);
			return new DidCommOutboundMessageContext(message, {
				agentContext: messageContext.agentContext,
				connection,
				associatedRecord: mediationRecord
			});
		}
	}
};

//#endregion
export { DidCommMediationRequestHandler };
