// @credo-ts/core/build/modules/dids/domain/key-type/ed25519.mjs


import { CredoError } from "../../../../error/CredoError.mjs";
import "../../../../error/index.mjs";
import { Ed25519PublicJwk } from "../../../kms/jwk/kty/okp/Ed25519PublicJwk.mjs";
import "../../../kms/index.mjs";
import { VERIFICATION_METHOD_TYPE_ED25519_VERIFICATION_KEY_2018, getEd25519VerificationKey2018, getPublicJwkFromEd25519VerificationKey2018, isEd25519VerificationKey2018 } from "../verificationMethod/Ed25519VerificationKey2018.mjs";
import { VERIFICATION_METHOD_TYPE_ED25519_VERIFICATION_KEY_2020, getPublicJwkFromEd25519VerificationKey2020, isEd25519VerificationKey2020 } from "../verificationMethod/Ed25519VerificationKey2020.mjs";
import { VERIFICATION_METHOD_TYPE_JSON_WEB_KEY_2020 } from "../verificationMethod/JsonWebKey2020.mjs";
import { VERIFICATION_METHOD_TYPE_MULTIKEY } from "../verificationMethod/Multikey.mjs";
import "../verificationMethod/index.mjs";
import { convertPublicKeyToX25519 as convertPublicKeyToX25519$1 } from "@stablelib/ed25519";

//#region src/modules/dids/domain/key-type/ed25519.ts
const keyDidEd25519 = {
	PublicJwkTypes: [Ed25519PublicJwk],
	supportedVerificationMethodTypes: [
		VERIFICATION_METHOD_TYPE_ED25519_VERIFICATION_KEY_2018,
		VERIFICATION_METHOD_TYPE_ED25519_VERIFICATION_KEY_2020,
		VERIFICATION_METHOD_TYPE_JSON_WEB_KEY_2020,
		VERIFICATION_METHOD_TYPE_MULTIKEY
	],
	getVerificationMethods: (did, publicJwk) => [getEd25519VerificationKey2018({
		id: `${did}#${publicJwk.fingerprint}`,
		publicJwk,
		controller: did
	})],
	getPublicJwkFromVerificationMethod: (verificationMethod) => {
		if (isEd25519VerificationKey2018(verificationMethod)) return getPublicJwkFromEd25519VerificationKey2018(verificationMethod);
		if (isEd25519VerificationKey2020(verificationMethod)) return getPublicJwkFromEd25519VerificationKey2020(verificationMethod);
		throw new CredoError(`Verification method with type '${verificationMethod.type}' not supported for key type Ed25519`);
	}
};

//#endregion
export { convertPublicKeyToX25519$1 as convertPublicKeyToX25519, keyDidEd25519 };
