// @credo-ts/didcomm/build/modules/credentials/protocol/v2/handlers/DidCommRequestCredentialV2Handler.mjs
import { getOutboundDidCommMessageContext } from "../../../../../getDidCommOutboundMessageContext.mjs";
import { DidCommRequestCredentialV2Message } from "../messages/DidCommRequestCredentialV2Message.mjs";
import { CredoError } from "@credo-ts/core";

//#region src/modules/credentials/protocol/v2/handlers/DidCommRequestCredentialV2Handler.ts
var DidCommRequestCredentialV2Handler = class {
	constructor(credentialProtocol) {
		this.supportedMessages = [DidCommRequestCredentialV2Message];
		this.credentialProtocol = credentialProtocol;
	}
	async handle(messageContext) {
		const credentialExchangeRecord = await this.credentialProtocol.processRequest(messageContext);
		if (await this.credentialProtocol.shouldAutoRespondToRequest(messageContext.agentContext, {
			credentialExchangeRecord,
			requestMessage: messageContext.message
		})) return await this.acceptRequest(credentialExchangeRecord, messageContext);
	}
	async acceptRequest(credentialExchangeRecord, messageContext) {
		messageContext.agentContext.config.logger.info("Automatically sending credential with autoAccept");
		const offerMessage = await this.credentialProtocol.findOfferMessage(messageContext.agentContext, credentialExchangeRecord.id);
		if (!offerMessage) throw new CredoError(`Could not find offer message for credential record with id ${credentialExchangeRecord.id}`);
		const { message } = await this.credentialProtocol.acceptRequest(messageContext.agentContext, { credentialExchangeRecord });
		return getOutboundDidCommMessageContext(messageContext.agentContext, {
			connectionRecord: messageContext.connection,
			message,
			associatedRecord: credentialExchangeRecord,
			lastReceivedMessage: messageContext.message,
			lastSentMessage: offerMessage
		});
	}
};

//#endregion
export { DidCommRequestCredentialV2Handler };
