// @credo-ts/didcomm/build/modules/credentials/protocol/v2/handlers/DidCommOfferCredentialV2Handler.mjs
import { getOutboundDidCommMessageContext } from "../../../../../getDidCommOutboundMessageContext.mjs";
import { DidCommOfferCredentialV2Message } from "../messages/DidCommOfferCredentialV2Message.mjs";

//#region src/modules/credentials/protocol/v2/handlers/DidCommOfferCredentialV2Handler.ts
var DidCommOfferCredentialV2Handler = class {
	constructor(credentialProtocol) {
		this.supportedMessages = [DidCommOfferCredentialV2Message];
		this.credentialProtocol = credentialProtocol;
	}
	async handle(messageContext) {
		const credentialExchangeRecord = await this.credentialProtocol.processOffer(messageContext);
		if (await this.credentialProtocol.shouldAutoRespondToOffer(messageContext.agentContext, {
			credentialExchangeRecord,
			offerMessage: messageContext.message
		})) return await this.acceptOffer(credentialExchangeRecord, messageContext);
	}
	async acceptOffer(credentialExchangeRecord, messageContext) {
		messageContext.agentContext.config.logger.info("Automatically sending request with autoAccept");
		const { message } = await this.credentialProtocol.acceptOffer(messageContext.agentContext, { credentialExchangeRecord });
		return getOutboundDidCommMessageContext(messageContext.agentContext, {
			connectionRecord: messageContext.connection,
			message,
			associatedRecord: credentialExchangeRecord,
			lastReceivedMessage: messageContext.message
		});
	}
};

//#endregion
export { DidCommOfferCredentialV2Handler };
