// @credo-ts/didcomm/build/modules/credentials/protocol/v2/handlers/DidCommIssueCredentialV2Handler.mjs
import { getOutboundDidCommMessageContext } from "../../../../../getDidCommOutboundMessageContext.mjs";
import { DidCommIssueCredentialV2Message } from "../messages/DidCommIssueCredentialV2Message.mjs";
import { CredoError } from "@credo-ts/core";

//#region src/modules/credentials/protocol/v2/handlers/DidCommIssueCredentialV2Handler.ts
var DidCommIssueCredentialV2Handler = class {
	constructor(credentialProtocol) {
		this.supportedMessages = [DidCommIssueCredentialV2Message];
		this.credentialProtocol = credentialProtocol;
	}
	async handle(messageContext) {
		const credentialExchangeRecord = await this.credentialProtocol.processCredential(messageContext);
		if (await this.credentialProtocol.shouldAutoRespondToCredential(messageContext.agentContext, {
			credentialExchangeRecord,
			credentialMessage: messageContext.message
		})) return await this.acceptCredential(credentialExchangeRecord, messageContext);
	}
	async acceptCredential(credentialExchangeRecord, messageContext) {
		messageContext.agentContext.config.logger.info("Automatically sending acknowledgement with autoAccept");
		const { message } = await this.credentialProtocol.acceptCredential(messageContext.agentContext, { credentialExchangeRecord });
		const requestMessage = await this.credentialProtocol.findRequestMessage(messageContext.agentContext, credentialExchangeRecord.id);
		if (!requestMessage) throw new CredoError(`No request message found for credential record with id '${credentialExchangeRecord.id}'`);
		return getOutboundDidCommMessageContext(messageContext.agentContext, {
			connectionRecord: messageContext.connection,
			message,
			associatedRecord: credentialExchangeRecord,
			lastReceivedMessage: messageContext.message,
			lastSentMessage: requestMessage
		});
	}
};

//#endregion
export { DidCommIssueCredentialV2Handler };
