// @credo-ts/didcomm/build/modules/credentials/protocol/v2/handlers/DidCommProposeCredentialV2Handler.mjs
import { DidCommOutboundMessageContext } from "../../../../../models/DidCommOutboundMessageContext.mjs";
import "../../../../../models/index.mjs";
import { DidCommProposeCredentialV2Message } from "../messages/DidCommProposeCredentialV2Message.mjs";

//#region src/modules/credentials/protocol/v2/handlers/DidCommProposeCredentialV2Handler.ts
var DidCommProposeCredentialV2Handler = class {
	constructor(credentialProtocol) {
		this.supportedMessages = [DidCommProposeCredentialV2Message];
		this.credentialProtocol = credentialProtocol;
	}
	async handle(messageContext) {
		const credentialExchangeRecord = await this.credentialProtocol.processProposal(messageContext);
		if (await this.credentialProtocol.shouldAutoRespondToProposal(messageContext.agentContext, {
			credentialExchangeRecord,
			proposalMessage: messageContext.message
		})) return await this.acceptProposal(credentialExchangeRecord, messageContext);
	}
	async acceptProposal(credentialExchangeRecord, messageContext) {
		messageContext.agentContext.config.logger.info("Automatically sending offer with autoAccept");
		if (!messageContext.connection) {
			messageContext.agentContext.config.logger.error("No connection on the messageContext, aborting auto accept");
			return;
		}
		const { message } = await this.credentialProtocol.acceptProposal(messageContext.agentContext, { credentialExchangeRecord });
		return new DidCommOutboundMessageContext(message, {
			agentContext: messageContext.agentContext,
			connection: messageContext.connection,
			associatedRecord: credentialExchangeRecord
		});
	}
};

//#endregion
export { DidCommProposeCredentialV2Handler };
