// @animo-id/pex/dist/module/lib/types/PexCredentialMapper.js
/* eslint-disable @typescript-eslint/no-explicit-any */
import { cborEncode, DateOnly, IssuerSignedDocument, MDoc, parseDeviceResponse, parseIssuerSigned, uint8ArrayToBase64Url } from '@animo-id/mdoc';
import { CredentialMapper, OriginalType, } from '@sphereon/ssi-types';
import * as u8a from 'uint8arrays';
// NOTE: this is partial reimplementation of CredentialMapper from ssi-types to overwrite the
// methods relying on Sphereon's mdoc implementation
export class PexCredentialMapper {
    static isMsoMdocDecodedPresentation(original) {
        return original instanceof MDoc;
    }
    static isMsoMdocDecodedCredential(original) {
        return original instanceof IssuerSignedDocument;
    }
    /**
     * Decodes a Verifiable Presentation to a uniform format.
     *
     * When decoding SD-JWT credentials, a hasher implementation must be provided. The hasher implementation must be sync. When using
     * an async hasher implementation, use the decodeSdJwtVcAsync method instead and you can provide the decoded payload to methods
     * instead of the compact SD-JWT.
     *
     * @param presentation
     * @param hasher Hasher implementation to use for SD-JWT decoding.
     */
    static decodeVerifiablePresentation(presentation, hasher) {
        if (CredentialMapper.isMsoMdocOid4VPEncoded(presentation)) {
            return presentation;
        }
        else if (this.isMsoMdocDecodedPresentation(presentation)) {
            return presentation;
        }
        else {
            return CredentialMapper.decodeVerifiablePresentation(presentation, hasher);
        }
    }
    static isW3cCredential(credential) {
        return CredentialMapper.isW3cCredential(credential);
    }
    /**
     * Converts a presentation to a wrapped presentation.
     *
     * When decoding SD-JWT credentials, a hasher implementation must be provided. The hasher implementation must be sync. When using
     * an async hasher implementation, use the decodeSdJwtVcAsync method instead and you can provide the decoded payload to methods
     * instead of the compact SD-JWT.
     *
     * @param hasher Hasher implementation to use for SD-JWT decoding
     */
    static toWrappedVerifiablePresentation(originalPresentation, opts) {
        // MSO_MDOC
        if (this.isMsoMdocDecodedPresentation(originalPresentation) || CredentialMapper.isMsoMdocOid4VPEncoded(originalPresentation)) {
            let deviceResponse;
            let originalType;
            if (CredentialMapper.isMsoMdocOid4VPEncoded(originalPresentation)) {
                deviceResponse = parseDeviceResponse(convertBase64urlToBinary(originalPresentation));
                originalType = OriginalType.MSO_MDOC_ENCODED;
            }
            else {
                deviceResponse = originalPresentation;
                originalType = OriginalType.MSO_MDOC_DECODED;
            }
            const mdocCredentials = deviceResponse.documents?.map((doc) => PexCredentialMapper.toWrappedVerifiableCredential(doc, opts));
            if (!mdocCredentials || mdocCredentials.length === 0) {
                throw new Error('could not extract any mdoc credentials from mdoc device response');
            }
            return {
                type: originalType,
                format: 'mso_mdoc',
                original: originalPresentation,
                presentation: deviceResponse,
                decoded: deviceResponse,
                vcs: mdocCredentials,
            };
        }
        return CredentialMapper.toWrappedVerifiablePresentation(originalPresentation, opts);
    }
    /**
     * Converts a credential to a wrapped credential.
     *
     * When decoding SD-JWT credentials, a hasher implementation must be provided. The hasher implementation must be sync. When using
     * an async hasher implementation, use the decodeSdJwtVcAsync method instead and you can provide the decoded payload to methods
     * instead of the compact SD-JWT.
     *
     * @param hasher Hasher implementation to use for SD-JWT decoding
     */
    static toWrappedVerifiableCredential(verifiableCredential, opts) {
        // MSO_MDOC
        if (this.isMsoMdocDecodedCredential(verifiableCredential) || CredentialMapper.isMsoMdocOid4VPEncoded(verifiableCredential)) {
            let mdoc;
            if (CredentialMapper.isMsoMdocOid4VPEncoded(verifiableCredential)) {
                mdoc = parseIssuerSigned(convertBase64urlToBinary(verifiableCredential));
            }
            else {
                mdoc = verifiableCredential;
            }
            return {
                type: CredentialMapper.isMsoMdocDecodedCredential(verifiableCredential)
                    ? OriginalType.MSO_MDOC_DECODED
                    : OriginalType.MSO_MDOC_ENCODED,
                format: 'mso_mdoc',
                original: verifiableCredential,
                credential: mdoc,
                decoded: getMdocDecodedPayload(mdoc),
            };
        }
        return CredentialMapper.toWrappedVerifiableCredential(verifiableCredential, opts);
    }
    static isW3cPresentation(presentation) {
        return CredentialMapper.isW3cPresentation(presentation);
    }
    static isWrappedSdJwtVerifiableCredential = (vc) => CredentialMapper.isWrappedSdJwtVerifiableCredential(vc);
    static isWrappedSdJwtVerifiablePresentation = (vc) => CredentialMapper.isWrappedSdJwtVerifiablePresentation(vc);
    static isWrappedW3CVerifiableCredential = (vc) => CredentialMapper.isWrappedW3CVerifiableCredential(vc);
    static isWrappedW3CVerifiablePresentation = (vc) => CredentialMapper.isWrappedW3CVerifiablePresentation(vc);
    static isWrappedMdocCredential = (vc) => CredentialMapper.isWrappedMdocCredential(vc);
    static isWrappedMdocPresentation = (vc) => CredentialMapper.isWrappedMdocPresentation(vc);
    static isSdJwtDecodedCredential(original) {
        return CredentialMapper.isSdJwtDecodedCredential(original);
    }
    static isJwtDecodedCredential(original) {
        return CredentialMapper.isJwtDecodedCredential(original);
    }
    static isSdJwtEncoded(original) {
        return CredentialMapper.isSdJwtEncoded(original);
    }
    static isJwtEncoded(original) {
        return CredentialMapper.isJwtEncoded(original);
    }
    static decodeVerifiableCredential(credential, hasher) {
        return CredentialMapper.decodeVerifiableCredential(credential, hasher);
    }
    static isCredential(original) {
        return CredentialMapper.isCredential(original);
    }
    static areOriginalVerifiableCredentialsEqual(firstOriginal, secondOriginal) {
        if (this.isMsoMdocDecodedCredential(firstOriginal) || this.isMsoMdocDecodedCredential(secondOriginal)) {
            return uint8ArrayToBase64Url(cborEncode(firstOriginal)) === uint8ArrayToBase64Url(cborEncode(secondOriginal));
        }
        return CredentialMapper.areOriginalVerifiableCredentialsEqual(firstOriginal, secondOriginal);
    }
}
export function getMdocDecodedPayload(mdoc) {
    const namespaces = mdoc.issuerSigned.nameSpaces;
    const decodedPayload = {};
    for (const [namespace, items] of Array.from(namespaces.entries())) {
        decodedPayload[namespace] = items.reduce((acc, item) => ({
            ...acc,
            [item.elementIdentifier]: encodeMdocValue(item.elementValue),
        }), {});
    }
    return decodedPayload;
}
function encodeMdocValue(value) {
    if (typeof value === 'string' || typeof value === 'number' || typeof value === 'undefined' || value === null)
        return value;
    if (value instanceof Date || value instanceof DateOnly)
        return value.toISOString();
    // TODO: we don't want undefined, but empty object might also not work?
    return {};
}
function convertBase64urlToBinary(data) {
    return u8a.fromString(data, 'base64url');
}
