// @credo-ts/didcomm/build/modules/connections/DidCommConnectionsModule.mjs
import { DidCommFeatureRegistry } from "../../DidCommFeatureRegistry.mjs";
import { DidCommMessageHandlerRegistry } from "../../DidCommMessageHandlerRegistry.mjs";
import { DidCommConnectionRole } from "./models/DidCommConnectionRole.mjs";
import { DidCommDidExchangeRole } from "./models/DidCommDidExchangeRole.mjs";
import { DidCommDidRotateRole } from "./models/DidCommDidRotateRole.mjs";
import "./models/index.mjs";
import { DidCommProtocol } from "../../models/features/DidCommProtocol.mjs";
import "../../models/index.mjs";
import { DidCommOutOfBandService } from "../oob/DidCommOutOfBandService.mjs";
import { DidCommConnectionRepository } from "./repository/DidCommConnectionRepository.mjs";
import "./repository/index.mjs";
import { DidCommConnectionService } from "./services/DidCommConnectionService.mjs";
import { DidCommConnectionsModuleConfig } from "./DidCommConnectionsModuleConfig.mjs";
import { DidCommDidRotateService } from "./services/DidCommDidRotateService.mjs";
import { DidCommTrustPingService } from "./services/DidCommTrustPingService.mjs";
import { DidCommRoutingService } from "../routing/services/DidCommRoutingService.mjs";
import { DidExchangeProtocol } from "./DidExchangeProtocol.mjs";
import { DidCommConnectionsApi } from "./DidCommConnectionsApi.mjs";
import { DidCommAckMessageHandler } from "./handlers/DidCommAckMessageHandler.mjs";
import { DidCommConnectionProblemReportHandler } from "./handlers/DidCommConnectionProblemReportHandler.mjs";
import { DidCommConnectionRequestHandler } from "./handlers/DidCommConnectionRequestHandler.mjs";
import { DidCommConnectionResponseHandler } from "./handlers/DidCommConnectionResponseHandler.mjs";
import { DidCommDidExchangeCompleteHandler } from "./handlers/DidCommDidExchangeCompleteHandler.mjs";
import { DidCommDidExchangeRequestHandler } from "./handlers/DidCommDidExchangeRequestHandler.mjs";
import { DidCommDidExchangeResponseHandler } from "./handlers/DidCommDidExchangeResponseHandler.mjs";
import { DidCommDidRotateAckHandler } from "./handlers/DidCommDidRotateAckHandler.mjs";
import { DidCommDidRotateHandler } from "./handlers/DidCommDidRotateHandler.mjs";
import { DidCommDidRotateProblemReportHandler } from "./handlers/DidCommDidRotateProblemReportHandler.mjs";
import { DidCommHangupHandler } from "./handlers/DidCommHangupHandler.mjs";
import { DidCommTrustPingMessageHandler } from "./handlers/DidCommTrustPingMessageHandler.mjs";
import { DidCommTrustPingResponseMessageHandler } from "./handlers/DidCommTrustPingResponseMessageHandler.mjs";
import "./handlers/index.mjs";
import { DidRepository, DidResolverService } from "@credo-ts/core";

//#region src/modules/connections/DidCommConnectionsModule.ts
var DidCommConnectionsModule = class {
	constructor(config) {
		this.api = DidCommConnectionsApi;
		this.config = new DidCommConnectionsModuleConfig(config);
	}
	/**
	* Registers the dependencies of the connections module on the dependency manager.
	*/
	register(dependencyManager) {
		dependencyManager.registerInstance(DidCommConnectionsModuleConfig, this.config);
		dependencyManager.registerSingleton(DidCommConnectionService);
		dependencyManager.registerSingleton(DidExchangeProtocol);
		dependencyManager.registerSingleton(DidCommDidRotateService);
		dependencyManager.registerSingleton(DidCommTrustPingService);
		dependencyManager.registerSingleton(DidCommConnectionRepository);
	}
	async initialize(agentContext) {
		const featureRegistry = agentContext.resolve(DidCommFeatureRegistry);
		const messageHandlerRegistry = agentContext.resolve(DidCommMessageHandlerRegistry);
		const connectionService = agentContext.resolve(DidCommConnectionService);
		const outOfBandService = agentContext.resolve(DidCommOutOfBandService);
		const routingService = agentContext.resolve(DidCommRoutingService);
		const didRepository = agentContext.resolve(DidRepository);
		const didResolverService = agentContext.resolve(DidResolverService);
		const trustPingService = agentContext.resolve(DidCommTrustPingService);
		const didExchangeProtocol = agentContext.resolve(DidExchangeProtocol);
		const didRotateService = agentContext.resolve(DidCommDidRotateService);
		messageHandlerRegistry.registerMessageHandler(new DidCommConnectionRequestHandler(connectionService, outOfBandService, routingService, didRepository, this.config));
		messageHandlerRegistry.registerMessageHandler(new DidCommConnectionResponseHandler(connectionService, outOfBandService, didResolverService, this.config));
		messageHandlerRegistry.registerMessageHandler(new DidCommAckMessageHandler(connectionService));
		messageHandlerRegistry.registerMessageHandler(new DidCommConnectionProblemReportHandler(connectionService));
		messageHandlerRegistry.registerMessageHandler(new DidCommTrustPingMessageHandler(trustPingService, connectionService));
		messageHandlerRegistry.registerMessageHandler(new DidCommTrustPingResponseMessageHandler(trustPingService));
		messageHandlerRegistry.registerMessageHandler(new DidCommDidExchangeRequestHandler(didExchangeProtocol, outOfBandService, routingService, didRepository, this.config));
		messageHandlerRegistry.registerMessageHandler(new DidCommDidExchangeResponseHandler(didExchangeProtocol, outOfBandService, connectionService, didResolverService, this.config));
		messageHandlerRegistry.registerMessageHandler(new DidCommDidExchangeCompleteHandler(didExchangeProtocol, outOfBandService));
		messageHandlerRegistry.registerMessageHandler(new DidCommDidRotateHandler(didRotateService));
		messageHandlerRegistry.registerMessageHandler(new DidCommDidRotateAckHandler(didRotateService));
		messageHandlerRegistry.registerMessageHandler(new DidCommHangupHandler(didRotateService));
		messageHandlerRegistry.registerMessageHandler(new DidCommDidRotateProblemReportHandler(didRotateService));
		featureRegistry.register(new DidCommProtocol({
			id: "https://didcomm.org/connections/1.0",
			roles: [DidCommConnectionRole.Invitee, DidCommConnectionRole.Inviter]
		}), new DidCommProtocol({
			id: "https://didcomm.org/didexchange/1.1",
			roles: [DidCommDidExchangeRole.Requester, DidCommDidExchangeRole.Responder]
		}), new DidCommProtocol({
			id: "https://didcomm.org/did-rotate/1.0",
			roles: [DidCommDidRotateRole.RotatingParty, DidCommDidRotateRole.ObservingParty]
		}));
	}
};

//#endregion
export { DidCommConnectionsModule };
