// @credo-ts/didcomm/build/modules/credentials/DidCommCredentialsModule.mjs
import { DidCommFeatureRegistry } from "../../DidCommFeatureRegistry.mjs";
import { DidCommMessageHandlerRegistry } from "../../DidCommMessageHandlerRegistry.mjs";
import { DidCommProtocol } from "../../models/features/DidCommProtocol.mjs";
import "../../models/index.mjs";
import { DidCommCredentialsModuleConfig } from "./DidCommCredentialsModuleConfig.mjs";
import { DidCommCredentialExchangeRepository } from "./repository/DidCommCredentialExchangeRepository.mjs";
import "./repository/index.mjs";
import { DidCommRevocationNotificationService } from "./protocol/revocation-notification/services/DidCommRevocationNotificationService.mjs";
import "./protocol/revocation-notification/services/index.mjs";
import { DidCommCredentialsApi } from "./DidCommCredentialsApi.mjs";
import { DidCommRevocationNotificationV1Handler } from "./protocol/revocation-notification/handlers/DidCommRevocationNotificationV1Handler.mjs";
import { DidCommRevocationNotificationV2Handler } from "./protocol/revocation-notification/handlers/DidCommRevocationNotificationV2Handler.mjs";
import "./protocol/revocation-notification/handlers/index.mjs";
import { DidCommCredentialV2Protocol } from "./protocol/v2/DidCommCredentialV2Protocol.mjs";
import "./protocol/v2/index.mjs";

//#region src/modules/credentials/DidCommCredentialsModule.ts
var DidCommCredentialsModule = class {
	constructor(config) {
		this.api = DidCommCredentialsApi;
		this.config = new DidCommCredentialsModuleConfig({
			...config,
			credentialProtocols: config?.credentialProtocols ?? [new DidCommCredentialV2Protocol({ credentialFormats: [] })]
		});
	}
	/**
	* Registers the dependencies of the credentials module on the dependency manager.
	*/
	register(dependencyManager) {
		dependencyManager.registerInstance(DidCommCredentialsModuleConfig, this.config);
		dependencyManager.registerSingleton(DidCommRevocationNotificationService);
		dependencyManager.registerSingleton(DidCommCredentialExchangeRepository);
	}
	async initialize(agentContext) {
		const messageHandlerRegistry = agentContext.dependencyManager.resolve(DidCommMessageHandlerRegistry);
		const featureRegistry = agentContext.dependencyManager.resolve(DidCommFeatureRegistry);
		const revocationNotificationService = agentContext.resolve(DidCommRevocationNotificationService);
		messageHandlerRegistry.registerMessageHandler(new DidCommRevocationNotificationV1Handler(revocationNotificationService));
		messageHandlerRegistry.registerMessageHandler(new DidCommRevocationNotificationV2Handler(revocationNotificationService));
		featureRegistry.register(new DidCommProtocol({
			id: "https://didcomm.org/revocation_notification/1.0",
			roles: ["holder"]
		}), new DidCommProtocol({
			id: "https://didcomm.org/revocation_notification/2.0",
			roles: ["holder"]
		}));
		for (const credentialProtocol of this.config.credentialProtocols) {
			credentialProtocol.register(messageHandlerRegistry, featureRegistry);
			if (credentialProtocol.constructor.name === "DidCommProofV1Protocol") agentContext.config.logger.debug("The 'DidCommProofV1Protocol' is deprecated and will be removed in version 0.7 of Credo. You should upgrade to the 'DidCommProofV2Protocol' instead.");
		}
	}
};

//#endregion
export { DidCommCredentialsModule };
