// @credo-ts/didcomm/build/messages/problem-reports/DidCommProblemReportMessage.mjs
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorate.mjs";
import { IsValidMessageType, parseMessageType } from "../../util/messageType.mjs";
import { __decorateMetadata } from "../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorateMetadata.mjs";
import { DidCommMessage } from "../../DidCommMessage.mjs";
import { IsEnum, IsOptional, IsString } from "class-validator";
import { Expose } from "class-transformer";

//#region src/messages/problem-reports/DidCommProblemReportMessage.ts
var _ref, _ref2, _ref3;
let WhoRetriesStatus = /* @__PURE__ */ function(WhoRetriesStatus) {
	WhoRetriesStatus["You"] = "YOU";
	WhoRetriesStatus["Me"] = "ME";
	WhoRetriesStatus["Both"] = "BOTH";
	WhoRetriesStatus["None"] = "NONE";
	return WhoRetriesStatus;
}({});
let ImpactStatus = /* @__PURE__ */ function(ImpactStatus) {
	ImpactStatus["Message"] = "MESSAGE";
	ImpactStatus["Thread"] = "THREAD";
	ImpactStatus["Connection"] = "CONNECTION";
	return ImpactStatus;
}({});
let WhereStatus = /* @__PURE__ */ function(WhereStatus) {
	WhereStatus["Cloud"] = "CLOUD";
	WhereStatus["Edge"] = "EDGE";
	WhereStatus["Wire"] = "WIRE";
	WhereStatus["Agency"] = "AGENCY";
	return WhereStatus;
}({});
let OtherStatus = /* @__PURE__ */ function(OtherStatus) {
	OtherStatus["You"] = "YOU";
	OtherStatus["Me"] = "ME";
	OtherStatus["Other"] = "OTHER";
	return OtherStatus;
}({});
/**
* @see https://github.com/hyperledger/aries-rfcs/blob/main/features/0035-report-problem/README.md
*/
var DidCommProblemReportMessage = class DidCommProblemReportMessage extends DidCommMessage {
	/**
	* Create new ReportProblem instance.
	* @param options
	*/
	constructor(options) {
		super();
		this.type = DidCommProblemReportMessage.type.messageTypeUri;
		if (options) {
			this.id = options.id || this.generateId();
			this.description = options.description;
			this.problemItems = options.problemItems;
			this.whoRetries = options.whoRetries;
			this.fixHint = options.fixHint;
			this.impact = options.impact;
			this.where = options.where;
			this.noticedTime = options.noticedTime;
			this.trackingUri = options.trackingUri;
			this.escalationUri = options.escalationUri;
		}
	}
};
DidCommProblemReportMessage.type = parseMessageType("https://didcomm.org/notification/1.0/problem-report");
__decorate([IsValidMessageType(DidCommProblemReportMessage.type), __decorateMetadata("design:type", String)], DidCommProblemReportMessage.prototype, "type", void 0);
__decorate([
	IsOptional(),
	Expose({ name: "problem_items" }),
	__decorateMetadata("design:type", Array)
], DidCommProblemReportMessage.prototype, "problemItems", void 0);
__decorate([
	IsOptional(),
	IsEnum(WhoRetriesStatus),
	Expose({ name: "who_retries" }),
	__decorateMetadata("design:type", typeof (_ref = typeof WhoRetriesStatus !== "undefined" && WhoRetriesStatus) === "function" ? _ref : Object)
], DidCommProblemReportMessage.prototype, "whoRetries", void 0);
__decorate([
	IsOptional(),
	Expose({ name: "fix_hint" }),
	__decorateMetadata("design:type", Object)
], DidCommProblemReportMessage.prototype, "fixHint", void 0);
__decorate([
	IsOptional(),
	IsEnum(WhereStatus),
	__decorateMetadata("design:type", typeof (_ref2 = typeof WhereStatus !== "undefined" && WhereStatus) === "function" ? _ref2 : Object)
], DidCommProblemReportMessage.prototype, "where", void 0);
__decorate([
	IsOptional(),
	IsEnum(ImpactStatus),
	__decorateMetadata("design:type", typeof (_ref3 = typeof ImpactStatus !== "undefined" && ImpactStatus) === "function" ? _ref3 : Object)
], DidCommProblemReportMessage.prototype, "impact", void 0);
__decorate([
	IsOptional(),
	IsString(),
	Expose({ name: "noticed_time" }),
	__decorateMetadata("design:type", String)
], DidCommProblemReportMessage.prototype, "noticedTime", void 0);
__decorate([
	IsOptional(),
	IsString(),
	Expose({ name: "tracking_uri" }),
	__decorateMetadata("design:type", String)
], DidCommProblemReportMessage.prototype, "trackingUri", void 0);
__decorate([
	IsOptional(),
	IsString(),
	Expose({ name: "escalation_uri" }),
	__decorateMetadata("design:type", String)
], DidCommProblemReportMessage.prototype, "escalationUri", void 0);

//#endregion
export { DidCommProblemReportMessage, ImpactStatus, OtherStatus, WhereStatus, WhoRetriesStatus };
