// @credo-ts/core/build/modules/vc/constants.mjs


//#region src/modules/vc/constants.ts
const SECURITY_CONTEXT_V1_URL = "https://w3id.org/security/v1";
const SECURITY_CONTEXT_V2_URL = "https://w3id.org/security/v2";
const SECURITY_CONTEXT_V3_URL = "https://w3id.org/security/v3-unstable";
const SECURITY_CONTEXT_URL = SECURITY_CONTEXT_V2_URL;
const SECURITY_X25519_CONTEXT_URL = "https://w3id.org/security/suites/x25519-2019/v1";
const DID_V1_CONTEXT_URL = "https://www.w3.org/ns/did/v1";
const CREDENTIALS_CONTEXT_V1_URL = "https://www.w3.org/2018/credentials/v1";
const CREDENTIALS_CONTEXT_V2_URL = "https://www.w3.org/ns/credentials/v2";
const CREDENTIALS_ISSUER_URL = "https://www.w3.org/2018/credentials#issuer";
const SECURITY_PROOF_URL = "https://w3id.org/security#proof";
const SECURITY_SIGNATURE_URL = "https://w3id.org/security#signature";
const VERIFIABLE_CREDENTIAL_TYPE = "VerifiableCredential";
const ENVELOPED_VERIFIABLE_CREDENTIAL_TYPE = "EnvelopedVerifiableCredential";
const VERIFIABLE_PRESENTATION_TYPE = "VerifiablePresentation";
const ENVELOPED_VERIFIABLE_PRESENTATION_TYPE = "EnvelopedVerifiablePresentation";
const EXPANDED_TYPE_CREDENTIALS_CONTEXT_V1_VC_TYPE = "https://www.w3.org/2018/credentials#VerifiableCredential";
const SECURITY_JWS_CONTEXT_URL = "https://w3id.org/security/suites/jws-2020/v1";
const SECURITY_CONTEXT_SECP256k1_URL = "https://w3id.org/security/suites/secp256k1-2019/v1";

//#endregion
export { CREDENTIALS_CONTEXT_V1_URL, CREDENTIALS_CONTEXT_V2_URL, CREDENTIALS_ISSUER_URL, DID_V1_CONTEXT_URL, ENVELOPED_VERIFIABLE_CREDENTIAL_TYPE, ENVELOPED_VERIFIABLE_PRESENTATION_TYPE, EXPANDED_TYPE_CREDENTIALS_CONTEXT_V1_VC_TYPE, SECURITY_CONTEXT_SECP256k1_URL, SECURITY_CONTEXT_URL, SECURITY_CONTEXT_V1_URL, SECURITY_CONTEXT_V2_URL, SECURITY_CONTEXT_V3_URL, SECURITY_JWS_CONTEXT_URL, SECURITY_PROOF_URL, SECURITY_SIGNATURE_URL, SECURITY_X25519_CONTEXT_URL, VERIFIABLE_CREDENTIAL_TYPE, VERIFIABLE_PRESENTATION_TYPE };
