// @credo-ts/core/build/modules/vc/W3cV2CredentialsApi.mjs


import { AgentContext } from "../../agent/context/AgentContext.mjs";
import { injectable } from "../../plugins/index.mjs";
import { __decorateMetadata } from "../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorateMetadata.mjs";
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorate.mjs";
import "../../agent/index.mjs";
import { W3cV2CredentialService } from "./W3cV2CredentialService.mjs";

//#region src/modules/vc/W3cV2CredentialsApi.ts
var _ref, _ref2;
let W3cV2CredentialsApi = class W3cV2CredentialsApi {
	constructor(agentContext, w3cV2CredentialService) {
		this.agentContext = agentContext;
		this.w3cV2CredentialService = w3cV2CredentialService;
	}
	async store(options) {
		return this.w3cV2CredentialService.storeCredential(this.agentContext, options);
	}
	async deleteById(id) {
		return this.w3cV2CredentialService.removeCredentialRecord(this.agentContext, id);
	}
	async getAll() {
		return this.w3cV2CredentialService.getAllCredentialRecords(this.agentContext);
	}
	async getById(id) {
		return this.w3cV2CredentialService.getCredentialRecordById(this.agentContext, id);
	}
	async findAllByQuery(query, queryOptions) {
		return this.w3cV2CredentialService.findCredentialsByQuery(this.agentContext, query, queryOptions);
	}
	async signCredential(options) {
		return this.w3cV2CredentialService.signCredential(this.agentContext, options);
	}
	async verifyCredential(options) {
		return this.w3cV2CredentialService.verifyCredential(this.agentContext, options);
	}
	async signPresentation(options) {
		return this.w3cV2CredentialService.signPresentation(this.agentContext, options);
	}
	async verifyPresentation(options) {
		return this.w3cV2CredentialService.verifyPresentation(this.agentContext, options);
	}
};
W3cV2CredentialsApi = __decorate([injectable(), __decorateMetadata("design:paramtypes", [typeof (_ref = typeof AgentContext !== "undefined" && AgentContext) === "function" ? _ref : Object, typeof (_ref2 = typeof W3cV2CredentialService !== "undefined" && W3cV2CredentialService) === "function" ? _ref2 : Object])], W3cV2CredentialsApi);

//#endregion
export { W3cV2CredentialsApi };
