// @credo-ts/core/build/utils/JsonTransformer.mjs


import { ClassValidationError } from "../error/ClassValidationError.mjs";
import { MessageValidator } from "./MessageValidator.mjs";
import { instanceToInstance, instanceToPlain, plainToInstance } from "class-transformer";

//#region src/utils/JsonTransformer.ts
var JsonTransformer = class JsonTransformer {
	static toJSON(classInstance) {
		return instanceToPlain(classInstance, { exposeDefaultValues: true });
	}
	static fromJSON(json, cls, { validate = true } = {}) {
		const instance = plainToInstance(cls, json, { exposeDefaultValues: true });
		if (!validate) return instance;
		if (!instance) throw new ClassValidationError("Cannot validate instance of ", { classType: Object.getPrototypeOf(cls).name });
		MessageValidator.validateSync(instance);
		return instance;
	}
	static clone(classInstance) {
		return instanceToInstance(classInstance, {
			exposeDefaultValues: true,
			enableCircularCheck: true,
			enableImplicitConversion: true,
			ignoreDecorators: true
		});
	}
	static serialize(classInstance) {
		return JSON.stringify(JsonTransformer.toJSON(classInstance));
	}
	static deserialize(jsonString, cls, { validate = true } = {}) {
		return JsonTransformer.fromJSON(JSON.parse(jsonString), cls, { validate });
	}
};

//#endregion
export { JsonTransformer };
