// @credo-ts/core/build/modules/vc/data-integrity/models/DataIntegrityProof.mjs


import { IsUri } from "../../../../utils/validators.mjs";
import "../../../../utils/index.mjs";
import { __decorateMetadata } from "../../../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorateMetadata.mjs";
import { __decorate } from "../../../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorate.mjs";
import { IsEnum, IsOptional, IsString } from "class-validator";

//#region src/modules/vc/data-integrity/models/DataIntegrityProof.ts
/**
* Linked Data Proof
* @see https://w3c.github.io/vc-data-model/#proofs-signatures
*
* @class LinkedDataProof
*/
var DataIntegrityProof = class {
	constructor(options) {
		if (options) {
			this.type = options.type;
			this.cryptosuite = options.cryptosuite;
			this.verificationMethod = options.verificationMethod;
			this.proofPurpose = options.proofPurpose;
			this.domain = options.domain;
			this.challenge = options.challenge;
			this.nonce = options.nonce;
			this.created = options.created;
			this.expires = options.expires;
			this.proofValue = options.proofValue;
			this.previousProof = options.previousProof;
		}
	}
};
__decorate([
	IsString(),
	IsEnum(["DataIntegrityProof"]),
	__decorateMetadata("design:type", String)
], DataIntegrityProof.prototype, "type", void 0);
__decorate([IsString(), __decorateMetadata("design:type", String)], DataIntegrityProof.prototype, "cryptosuite", void 0);
__decorate([IsString(), __decorateMetadata("design:type", String)], DataIntegrityProof.prototype, "proofPurpose", void 0);
__decorate([IsString(), __decorateMetadata("design:type", String)], DataIntegrityProof.prototype, "verificationMethod", void 0);
__decorate([
	IsUri(),
	IsOptional(),
	__decorateMetadata("design:type", String)
], DataIntegrityProof.prototype, "domain", void 0);
__decorate([
	IsString(),
	IsOptional(),
	__decorateMetadata("design:type", String)
], DataIntegrityProof.prototype, "challenge", void 0);
__decorate([
	IsString(),
	IsOptional(),
	__decorateMetadata("design:type", String)
], DataIntegrityProof.prototype, "nonce", void 0);
__decorate([
	IsString(),
	IsOptional(),
	__decorateMetadata("design:type", String)
], DataIntegrityProof.prototype, "created", void 0);
__decorate([
	IsString(),
	IsOptional(),
	__decorateMetadata("design:type", String)
], DataIntegrityProof.prototype, "expires", void 0);
__decorate([
	IsString(),
	IsOptional(),
	__decorateMetadata("design:type", String)
], DataIntegrityProof.prototype, "proofValue", void 0);
__decorate([
	IsString(),
	IsOptional(),
	__decorateMetadata("design:type", String)
], DataIntegrityProof.prototype, "previousProof", void 0);

//#endregion
export { DataIntegrityProof };
