// @credo-ts/core/build/modules/vc/data-integrity/models/LinkedDataProof.mjs


import { IsUri } from "../../../../utils/validators.mjs";
import "../../../../utils/index.mjs";
import { __decorateMetadata } from "../../../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorateMetadata.mjs";
import { __decorate } from "../../../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorate.mjs";
import { IsOptional, IsString } from "class-validator";

//#region src/modules/vc/data-integrity/models/LinkedDataProof.ts
/**
* Linked Data Proof
* @see https://w3c.github.io/vc-data-model/#proofs-signatures
*
* @class LinkedDataProof
*/
var LinkedDataProof = class {
	constructor(options) {
		if (options) {
			this.type = options.type;
			this.proofPurpose = options.proofPurpose;
			this.verificationMethod = options.verificationMethod;
			this.created = options.created;
			this.domain = options.domain;
			this.challenge = options.challenge;
			this.jws = options.jws;
			this.proofValue = options.proofValue;
			this.nonce = options.nonce;
		}
	}
};
__decorate([IsString(), __decorateMetadata("design:type", String)], LinkedDataProof.prototype, "type", void 0);
__decorate([IsString(), __decorateMetadata("design:type", String)], LinkedDataProof.prototype, "proofPurpose", void 0);
__decorate([IsString(), __decorateMetadata("design:type", String)], LinkedDataProof.prototype, "verificationMethod", void 0);
__decorate([IsString(), __decorateMetadata("design:type", String)], LinkedDataProof.prototype, "created", void 0);
__decorate([
	IsUri(),
	IsOptional(),
	__decorateMetadata("design:type", String)
], LinkedDataProof.prototype, "domain", void 0);
__decorate([
	IsString(),
	IsOptional(),
	__decorateMetadata("design:type", String)
], LinkedDataProof.prototype, "challenge", void 0);
__decorate([
	IsString(),
	IsOptional(),
	__decorateMetadata("design:type", String)
], LinkedDataProof.prototype, "jws", void 0);
__decorate([
	IsString(),
	IsOptional(),
	__decorateMetadata("design:type", String)
], LinkedDataProof.prototype, "proofValue", void 0);
__decorate([
	IsString(),
	IsOptional(),
	__decorateMetadata("design:type", String)
], LinkedDataProof.prototype, "nonce", void 0);

//#endregion
export { LinkedDataProof };
