// @credo-ts/didcomm/build/modules/routing/DidCommMediationRecipientModuleConfig.mjs
//#region src/modules/routing/DidCommMediationRecipientModuleConfig.ts
var DidCommMediationRecipientModuleConfig = class {
	constructor(options) {
		this.options = options ?? {};
	}
	/** See {@link DidCommMediationRecipientModuleConfigOptions.mediatorPollingInterval} */
	get mediatorPollingInterval() {
		return this.options.mediatorPollingInterval ?? 5e3;
	}
	/** See {@link DidCommMediationRecipientModuleConfigOptions.mediatorPickupStrategy} */
	get mediatorPickupStrategy() {
		return this.options.mediatorPickupStrategy;
	}
	/** See {@link DidCommMediationRecipientModuleConfigOptions.maximumMessagePickup} */
	get maximumMessagePickup() {
		return this.options.maximumMessagePickup ?? 10;
	}
	/** See {@link DidCommMediationRecipientModuleConfigOptions.baseMediatorReconnectionIntervalMs} */
	get baseMediatorReconnectionIntervalMs() {
		return this.options.baseMediatorReconnectionIntervalMs ?? 100;
	}
	/** See {@link DidCommMediationRecipientModuleConfigOptions.maximumMediatorReconnectionIntervalMs} */
	get maximumMediatorReconnectionIntervalMs() {
		return this.options.maximumMediatorReconnectionIntervalMs ?? Number.POSITIVE_INFINITY;
	}
	/** See {@link DidCommMediationRecipientModuleConfigOptions.mediatorInvitationUrl} */
	get mediatorInvitationUrl() {
		return this.options.mediatorInvitationUrl;
	}
};

//#endregion
export { DidCommMediationRecipientModuleConfig };
