// @credo-ts/didcomm/build/modules/routing/DidCommMediatorModule.mjs
import { DidCommFeatureRegistry } from "../../DidCommFeatureRegistry.mjs";
import { DidCommMessageHandlerRegistry } from "../../DidCommMessageHandlerRegistry.mjs";
import { DidCommMediatorRoutingRepository } from "./repository/DidCommMediatorRoutingRepository.mjs";
import { DidCommProtocol } from "../../models/features/DidCommProtocol.mjs";
import "../../models/index.mjs";
import { DidCommMediationRole } from "./models/DidCommMediationRole.mjs";
import { DidCommMediationRepository } from "./repository/DidCommMediationRepository.mjs";
import "./repository/index.mjs";
import { DidCommForwardHandler } from "./handlers/DidCommForwardHandler.mjs";
import { DidCommKeylistUpdateHandler } from "./handlers/DidCommKeylistUpdateHandler.mjs";
import { DidCommMediationRequestHandler } from "./handlers/DidCommMediationRequestHandler.mjs";
import { DidCommMediatorModuleConfig } from "./DidCommMediatorModuleConfig.mjs";
import { DidCommMediatorService } from "./services/DidCommMediatorService.mjs";
import "./services/index.mjs";
import { DidCommMediatorApi } from "./DidCommMediatorApi.mjs";

//#region src/modules/routing/DidCommMediatorModule.ts
var DidCommMediatorModule = class {
	constructor(config) {
		this.api = DidCommMediatorApi;
		this.config = new DidCommMediatorModuleConfig(config);
	}
	/**
	* Registers the dependencies of the question answer module on the dependency manager.
	*/
	register(dependencyManager) {
		dependencyManager.registerInstance(DidCommMediatorModuleConfig, this.config);
		dependencyManager.registerSingleton(DidCommMediatorService);
		dependencyManager.registerSingleton(DidCommMediationRepository);
		dependencyManager.registerSingleton(DidCommMediatorRoutingRepository);
	}
	async initialize(agentContext) {
		const featureRegistry = agentContext.dependencyManager.resolve(DidCommFeatureRegistry);
		const messageHandlerRegistry = agentContext.resolve(DidCommMessageHandlerRegistry);
		const mediatorService = agentContext.resolve(DidCommMediatorService);
		messageHandlerRegistry.registerMessageHandler(new DidCommKeylistUpdateHandler(mediatorService));
		messageHandlerRegistry.registerMessageHandler(new DidCommForwardHandler(mediatorService));
		messageHandlerRegistry.registerMessageHandler(new DidCommMediationRequestHandler(mediatorService, this.config));
		featureRegistry.register(new DidCommProtocol({
			id: "https://didcomm.org/coordinate-mediation/1.0",
			roles: [DidCommMediationRole.Mediator]
		}));
	}
	async onInitializeContext(agentContext) {
		if (!agentContext.isRootAgentContext) return;
		const mediatorService = agentContext.dependencyManager.resolve(DidCommMediatorService);
		agentContext.config.logger.debug("Mediator routing record not loaded yet, retrieving from storage");
		if (!await mediatorService.findMediatorRoutingRecord(agentContext)) {
			agentContext.config.logger.debug("Mediator routing record does not exist yet, creating routing keys and record");
			await mediatorService.createMediatorRoutingRecord(agentContext);
		}
	}
};

//#endregion
export { DidCommMediatorModule };
