// @credo-ts/didcomm/build/modules/routing/DidCommMediationRecipientModule.mjs
import { DidCommFeatureRegistry } from "../../DidCommFeatureRegistry.mjs";
import { DidCommMessageHandlerRegistry } from "../../DidCommMessageHandlerRegistry.mjs";
import { DidCommProtocol } from "../../models/features/DidCommProtocol.mjs";
import "../../models/index.mjs";
import { DidCommMediationRole } from "./models/DidCommMediationRole.mjs";
import { DidCommMediationRepository } from "./repository/DidCommMediationRepository.mjs";
import { DidCommMediationRecipientService } from "./services/DidCommMediationRecipientService.mjs";
import { DidCommRoutingService } from "./services/DidCommRoutingService.mjs";
import { DidCommConnectionsApi } from "../connections/DidCommConnectionsApi.mjs";
import "../connections/index.mjs";
import { DidCommOutOfBandApi } from "../oob/DidCommOutOfBandApi.mjs";
import "../oob/index.mjs";
import { DidCommMediationRecipientModuleConfig } from "./DidCommMediationRecipientModuleConfig.mjs";
import "./repository/index.mjs";
import { DidCommMediationRecipientApi } from "./DidCommMediationRecipientApi.mjs";
import { DidCommKeylistUpdateResponseHandler } from "./handlers/DidCommKeylistUpdateResponseHandler.mjs";
import { DidCommMediationDenyHandler } from "./handlers/DidCommMediationDenyHandler.mjs";
import { DidCommMediationGrantHandler } from "./handlers/DidCommMediationGrantHandler.mjs";
import "./handlers/index.mjs";
import "./services/index.mjs";
import { CredoError } from "@credo-ts/core";

//#region src/modules/routing/DidCommMediationRecipientModule.ts
var DidCommMediationRecipientModule = class {
	constructor(config) {
		this.api = DidCommMediationRecipientApi;
		this.config = new DidCommMediationRecipientModuleConfig(config);
	}
	/**
	* Registers the dependencies of the mediator recipient module on the dependency manager.
	*/
	register(dependencyManager) {
		dependencyManager.registerInstance(DidCommMediationRecipientModuleConfig, this.config);
		dependencyManager.registerSingleton(DidCommMediationRecipientService);
		dependencyManager.registerSingleton(DidCommRoutingService);
		dependencyManager.registerSingleton(DidCommMediationRepository);
	}
	async initialize(agentContext) {
		const featureRegistry = agentContext.dependencyManager.resolve(DidCommFeatureRegistry);
		const messageHandlerRegistry = agentContext.resolve(DidCommMessageHandlerRegistry);
		const mediationRecipientService = agentContext.resolve(DidCommMediationRecipientService);
		messageHandlerRegistry.registerMessageHandler(new DidCommKeylistUpdateResponseHandler(mediationRecipientService));
		messageHandlerRegistry.registerMessageHandler(new DidCommMediationGrantHandler(mediationRecipientService));
		messageHandlerRegistry.registerMessageHandler(new DidCommMediationDenyHandler(mediationRecipientService));
		featureRegistry.register(new DidCommProtocol({
			id: "https://didcomm.org/coordinate-mediation/1.0",
			roles: [DidCommMediationRole.Recipient]
		}));
	}
	async onCloseContext(agentContext) {
		if (!agentContext.isRootAgentContext) return;
		await agentContext.dependencyManager.resolve(DidCommMediationRecipientApi).stopMessagePickup();
	}
	async onInitializeContext(agentContext) {
		if (!agentContext.isRootAgentContext) return;
		const mediationRecipientApi = agentContext.dependencyManager.resolve(DidCommMediationRecipientApi);
		if (this.config.mediatorInvitationUrl) {
			agentContext.config.logger.debug("Provision mediation with invitation", { mediatorInvitationUrl: this.config.mediatorInvitationUrl });
			const mediationConnection = await this.getMediationConnection(agentContext, this.config.mediatorInvitationUrl);
			await mediationRecipientApi.provision(mediationConnection);
		}
		const defaultMediator = await mediationRecipientApi.findDefaultMediator();
		if (defaultMediator) mediationRecipientApi.initiateMessagePickup(defaultMediator).catch((error) => {
			agentContext.config.logger.warn(`Error initiating message pickup with mediator ${defaultMediator.id}`, { error });
		});
	}
	async getMediationConnection(agentContext, mediatorInvitationUrl) {
		const oobApi = agentContext.dependencyManager.resolve(DidCommOutOfBandApi);
		const connectionsApi = agentContext.dependencyManager.resolve(DidCommConnectionsApi);
		const mediationRecipientApi = agentContext.dependencyManager.resolve(DidCommMediationRecipientApi);
		const outOfBandInvitation = await oobApi.parseInvitation(mediatorInvitationUrl);
		const outOfBandRecord = await oobApi.findByReceivedInvitationId(outOfBandInvitation.id);
		const [connection] = outOfBandRecord ? await connectionsApi.findAllByOutOfBandId(outOfBandRecord.id) : [];
		if (!connection) {
			agentContext.config.logger.debug("Mediation connection does not exist, creating connection");
			const routing = await mediationRecipientApi.getRouting({ useDefaultMediator: false });
			agentContext.config.logger.debug("Routing created", routing);
			const { connectionRecord: newConnection } = await oobApi.receiveInvitation(outOfBandInvitation, {
				label: "",
				routing
			});
			agentContext.config.logger.debug("Mediation invitation processed", { outOfBandInvitation });
			if (!newConnection) throw new CredoError("No connection record to provision mediation.");
			return connectionsApi.returnWhenIsConnected(newConnection.id);
		}
		if (!connection.isReady) return connectionsApi.returnWhenIsConnected(connection.id);
		return connection;
	}
};

//#endregion
export { DidCommMediationRecipientModule };
