// @credo-ts/didcomm/build/modules/credentials/formats/jsonld/DidCommJsonLdCredentialDetailOptions.mjs
import { __decorate } from "../../../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorate.mjs";
import { __decorateMetadata } from "../../../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorateMetadata.mjs";
import { IsObject, IsOptional, IsString } from "class-validator";

//#region src/modules/credentials/formats/jsonld/DidCommJsonLdCredentialDetailOptions.ts
var _ref;
var DidCommJsonLdCredentialDetailCredentialStatus = class {
	constructor(options) {
		if (options) this.type = options.type;
	}
};
__decorate([IsString(), __decorateMetadata("design:type", String)], DidCommJsonLdCredentialDetailCredentialStatus.prototype, "type", void 0);
var DidCommJsonLdCredentialDetailOptions = class {
	constructor(options) {
		if (options) {
			this.proofPurpose = options.proofPurpose;
			this.created = options.created;
			this.domain = options.domain;
			this.challenge = options.challenge;
			this.credentialStatus = options.credentialStatus;
			this.proofType = options.proofType;
		}
	}
};
__decorate([IsString(), __decorateMetadata("design:type", String)], DidCommJsonLdCredentialDetailOptions.prototype, "proofPurpose", void 0);
__decorate([
	IsString(),
	IsOptional(),
	__decorateMetadata("design:type", String)
], DidCommJsonLdCredentialDetailOptions.prototype, "created", void 0);
__decorate([
	IsString(),
	IsOptional(),
	__decorateMetadata("design:type", String)
], DidCommJsonLdCredentialDetailOptions.prototype, "domain", void 0);
__decorate([
	IsString(),
	IsOptional(),
	__decorateMetadata("design:type", String)
], DidCommJsonLdCredentialDetailOptions.prototype, "challenge", void 0);
__decorate([IsString(), __decorateMetadata("design:type", String)], DidCommJsonLdCredentialDetailOptions.prototype, "proofType", void 0);
__decorate([
	IsOptional(),
	IsObject(),
	__decorateMetadata("design:type", typeof (_ref = typeof DidCommJsonLdCredentialDetailCredentialStatus !== "undefined" && DidCommJsonLdCredentialDetailCredentialStatus) === "function" ? _ref : Object)
], DidCommJsonLdCredentialDetailOptions.prototype, "credentialStatus", void 0);

//#endregion
export { DidCommJsonLdCredentialDetailCredentialStatus, DidCommJsonLdCredentialDetailOptions };
