// @credo-ts/didcomm/build/modules/proofs/protocol/v2/DidCommProofFormatCoordinator.mjs
import { DidCommMessageRepository } from "../../../../repository/DidCommMessageRepository.mjs";
import { DidCommMessageRole } from "../../../../repository/DidCommMessageRole.mjs";
import "../../../../repository/index.mjs";
import { DidCommPresentationV2Message } from "./messages/DidCommPresentationV2Message.mjs";
import { DidCommProposePresentationV2Message } from "./messages/DidCommProposePresentationV2Message.mjs";
import { DidCommRequestPresentationV2Message } from "./messages/DidCommRequestPresentationV2Message.mjs";
import "./messages/index.mjs";
import { CredoError } from "@credo-ts/core";

//#region src/modules/proofs/protocol/v2/DidCommProofFormatCoordinator.ts
var DidCommProofFormatCoordinator = class {
	/**
	* Create a {@link DidCommProposePresentationV2Message}.
	*
	* @param options
	* @returns The created {@link DidCommProposePresentationV2Message}
	*
	*/
	async createProposal(agentContext, { proofFormats, formatServices, proofRecord, comment, goalCode, goal }) {
		const didCommMessageRepository = agentContext.dependencyManager.resolve(DidCommMessageRepository);
		const formats = [];
		const proposalAttachments = [];
		for (const formatService of formatServices) {
			const { format, attachment } = await formatService.createProposal(agentContext, {
				proofFormats,
				proofRecord
			});
			proposalAttachments.push(attachment);
			formats.push(format);
		}
		const message = new DidCommProposePresentationV2Message({
			id: proofRecord.threadId,
			formats,
			proposalAttachments,
			comment,
			goalCode,
			goal
		});
		message.setThread({
			threadId: proofRecord.threadId,
			parentThreadId: proofRecord.parentThreadId
		});
		await didCommMessageRepository.saveOrUpdateAgentMessage(agentContext, {
			agentMessage: message,
			role: DidCommMessageRole.Sender,
			associatedRecordId: proofRecord.id
		});
		return message;
	}
	async processProposal(agentContext, { proofRecord, message, formatServices }) {
		const didCommMessageRepository = agentContext.dependencyManager.resolve(DidCommMessageRepository);
		for (const formatService of formatServices) {
			const attachment = this.getAttachmentForService(formatService, message.formats, message.proposalAttachments);
			await formatService.processProposal(agentContext, {
				attachment,
				proofRecord
			});
		}
		await didCommMessageRepository.saveOrUpdateAgentMessage(agentContext, {
			agentMessage: message,
			role: DidCommMessageRole.Receiver,
			associatedRecordId: proofRecord.id
		});
	}
	async acceptProposal(agentContext, { proofRecord, proofFormats, formatServices, comment, goalCode, goal, presentMultiple, willConfirm }) {
		const didCommMessageRepository = agentContext.dependencyManager.resolve(DidCommMessageRepository);
		const formats = [];
		const requestAttachments = [];
		const proposalMessage = await didCommMessageRepository.getAgentMessage(agentContext, {
			associatedRecordId: proofRecord.id,
			messageClass: DidCommProposePresentationV2Message,
			role: DidCommMessageRole.Receiver
		});
		for (const formatService of formatServices) {
			const proposalAttachment = this.getAttachmentForService(formatService, proposalMessage.formats, proposalMessage.proposalAttachments);
			const { attachment, format } = await formatService.acceptProposal(agentContext, {
				proofRecord,
				proofFormats,
				proposalAttachment
			});
			requestAttachments.push(attachment);
			formats.push(format);
		}
		const message = new DidCommRequestPresentationV2Message({
			formats,
			requestAttachments,
			comment,
			goalCode,
			goal,
			presentMultiple,
			willConfirm
		});
		message.setThread({
			threadId: proofRecord.threadId,
			parentThreadId: proofRecord.parentThreadId
		});
		await didCommMessageRepository.saveOrUpdateAgentMessage(agentContext, {
			agentMessage: message,
			associatedRecordId: proofRecord.id,
			role: DidCommMessageRole.Sender
		});
		return message;
	}
	/**
	* Create a {@link DidCommRequestPresentationV2Message}.
	*
	* @param options
	* @returns The created {@link DidCommRequestPresentationV2Message}
	*
	*/
	async createRequest(agentContext, { proofFormats, formatServices, proofRecord, comment, goalCode, goal, presentMultiple, willConfirm }) {
		const didCommMessageRepository = agentContext.dependencyManager.resolve(DidCommMessageRepository);
		const formats = [];
		const requestAttachments = [];
		for (const formatService of formatServices) {
			const { format, attachment } = await formatService.createRequest(agentContext, {
				proofFormats,
				proofRecord
			});
			requestAttachments.push(attachment);
			formats.push(format);
		}
		const message = new DidCommRequestPresentationV2Message({
			formats,
			comment,
			requestAttachments,
			goalCode,
			goal,
			presentMultiple,
			willConfirm
		});
		message.setThread({
			threadId: proofRecord.threadId,
			parentThreadId: proofRecord.parentThreadId
		});
		await didCommMessageRepository.saveOrUpdateAgentMessage(agentContext, {
			agentMessage: message,
			role: DidCommMessageRole.Sender,
			associatedRecordId: proofRecord.id
		});
		return message;
	}
	async processRequest(agentContext, { proofRecord, message, formatServices }) {
		const didCommMessageRepository = agentContext.dependencyManager.resolve(DidCommMessageRepository);
		for (const formatService of formatServices) {
			const attachment = this.getAttachmentForService(formatService, message.formats, message.requestAttachments);
			await formatService.processRequest(agentContext, {
				attachment,
				proofRecord
			});
		}
		await didCommMessageRepository.saveOrUpdateAgentMessage(agentContext, {
			agentMessage: message,
			role: DidCommMessageRole.Receiver,
			associatedRecordId: proofRecord.id
		});
	}
	async acceptRequest(agentContext, { proofRecord, proofFormats, formatServices, comment, lastPresentation, goalCode, goal }) {
		const didCommMessageRepository = agentContext.dependencyManager.resolve(DidCommMessageRepository);
		const requestMessage = await didCommMessageRepository.getAgentMessage(agentContext, {
			associatedRecordId: proofRecord.id,
			messageClass: DidCommRequestPresentationV2Message,
			role: DidCommMessageRole.Receiver
		});
		const proposalMessage = await didCommMessageRepository.findAgentMessage(agentContext, {
			associatedRecordId: proofRecord.id,
			messageClass: DidCommProposePresentationV2Message,
			role: DidCommMessageRole.Receiver
		});
		const formats = [];
		const presentationAttachments = [];
		for (const formatService of formatServices) {
			const requestAttachment = this.getAttachmentForService(formatService, requestMessage.formats, requestMessage.requestAttachments);
			const proposalAttachment = proposalMessage ? this.getAttachmentForService(formatService, proposalMessage.formats, proposalMessage.proposalAttachments) : void 0;
			const { attachment, format } = await formatService.acceptRequest(agentContext, {
				requestAttachment,
				proposalAttachment,
				proofRecord,
				proofFormats
			});
			presentationAttachments.push(attachment);
			formats.push(format);
		}
		const message = new DidCommPresentationV2Message({
			formats,
			presentationAttachments,
			comment,
			lastPresentation,
			goalCode,
			goal
		});
		message.setThread({
			threadId: proofRecord.threadId,
			parentThreadId: proofRecord.parentThreadId
		});
		message.setPleaseAck();
		await didCommMessageRepository.saveOrUpdateAgentMessage(agentContext, {
			agentMessage: message,
			associatedRecordId: proofRecord.id,
			role: DidCommMessageRole.Sender
		});
		return message;
	}
	async getCredentialsForRequest(agentContext, { proofRecord, proofFormats, formatServices }) {
		const didCommMessageRepository = agentContext.dependencyManager.resolve(DidCommMessageRepository);
		const requestMessage = await didCommMessageRepository.getAgentMessage(agentContext, {
			associatedRecordId: proofRecord.id,
			messageClass: DidCommRequestPresentationV2Message,
			role: DidCommMessageRole.Receiver
		});
		const proposalMessage = await didCommMessageRepository.findAgentMessage(agentContext, {
			associatedRecordId: proofRecord.id,
			messageClass: DidCommProposePresentationV2Message,
			role: DidCommMessageRole.Receiver
		});
		const credentialsForRequest = {};
		for (const formatService of formatServices) {
			const requestAttachment = this.getAttachmentForService(formatService, requestMessage.formats, requestMessage.requestAttachments);
			const proposalAttachment = proposalMessage ? this.getAttachmentForService(formatService, proposalMessage.formats, proposalMessage.proposalAttachments) : void 0;
			const credentialsForFormat = await formatService.getCredentialsForRequest(agentContext, {
				requestAttachment,
				proposalAttachment,
				proofRecord,
				proofFormats
			});
			credentialsForRequest[formatService.formatKey] = credentialsForFormat;
		}
		return credentialsForRequest;
	}
	async selectCredentialsForRequest(agentContext, { proofRecord, proofFormats, formatServices }) {
		const didCommMessageRepository = agentContext.dependencyManager.resolve(DidCommMessageRepository);
		const requestMessage = await didCommMessageRepository.getAgentMessage(agentContext, {
			associatedRecordId: proofRecord.id,
			messageClass: DidCommRequestPresentationV2Message,
			role: DidCommMessageRole.Receiver
		});
		const proposalMessage = await didCommMessageRepository.findAgentMessage(agentContext, {
			associatedRecordId: proofRecord.id,
			messageClass: DidCommProposePresentationV2Message,
			role: DidCommMessageRole.Receiver
		});
		const credentialsForRequest = {};
		for (const formatService of formatServices) {
			const requestAttachment = this.getAttachmentForService(formatService, requestMessage.formats, requestMessage.requestAttachments);
			const proposalAttachment = proposalMessage ? this.getAttachmentForService(formatService, proposalMessage.formats, proposalMessage.proposalAttachments) : void 0;
			const credentialsForFormat = await formatService.selectCredentialsForRequest(agentContext, {
				requestAttachment,
				proposalAttachment,
				proofRecord,
				proofFormats
			});
			credentialsForRequest[formatService.formatKey] = credentialsForFormat;
		}
		return credentialsForRequest;
	}
	async processPresentation(agentContext, { proofRecord, message, requestMessage, formatServices }) {
		const didCommMessageRepository = agentContext.dependencyManager.resolve(DidCommMessageRepository);
		const formatVerificationResults = [];
		for (const formatService of formatServices) {
			const attachment = this.getAttachmentForService(formatService, message.formats, message.presentationAttachments);
			const requestAttachment = this.getAttachmentForService(formatService, requestMessage.formats, requestMessage.requestAttachments);
			try {
				const isValid = await formatService.processPresentation(agentContext, {
					attachment,
					requestAttachment,
					proofRecord
				});
				formatVerificationResults.push(isValid);
			} catch (error) {
				return {
					message: error.message,
					isValid: false
				};
			}
		}
		await didCommMessageRepository.saveOrUpdateAgentMessage(agentContext, {
			agentMessage: message,
			role: DidCommMessageRole.Receiver,
			associatedRecordId: proofRecord.id
		});
		const isValid = formatVerificationResults.every((isValid) => isValid === true);
		if (isValid) return {
			isValid,
			message: void 0
		};
		return {
			isValid,
			message: "Not all presentations are valid"
		};
	}
	getAttachmentForService(credentialFormatService, formats, attachments) {
		const attachmentId = this.getAttachmentIdForService(credentialFormatService, formats);
		const attachment = attachments.find((attachment) => attachment.id === attachmentId);
		if (!attachment) throw new CredoError(`DidCommAttachment with id ${attachmentId} not found in attachments.`);
		return attachment;
	}
	getAttachmentIdForService(credentialFormatService, formats) {
		const format = formats.find((format) => credentialFormatService.supportsFormat(format.format));
		if (!format) throw new CredoError(`No attachment found for service ${credentialFormatService.formatKey}`);
		return format.attachmentId;
	}
};

//#endregion
export { DidCommProofFormatCoordinator };
