// @animo-id/pex/dist/module/lib/evaluation/handlers/formatRestrictionEvaluationHandler.js
import { Status } from '../../ConstraintUtils';
import PexMessages from '../../types/Messages';
import { AbstractEvaluationHandler } from './abstractEvaluationHandler';
export class FormatRestrictionEvaluationHandler extends AbstractEvaluationHandler {
    constructor(client) {
        super(client);
    }
    getName() {
        return 'FormatRestrictionEvaluation';
    }
    handle(pd, wrappedVcs) {
        const restrictToFormats = this.client.restrictToFormats ? Object.keys(this.client.restrictToFormats) : undefined;
        pd.input_descriptors.forEach((_inputDescriptor, index) => {
            wrappedVcs.forEach((wvc, vcIndex) => {
                const formats = 'format' in _inputDescriptor && _inputDescriptor.format ? Object.keys(_inputDescriptor.format) : [wvc.format];
                let allowedFormats = restrictToFormats ?? formats;
                if ('format' in _inputDescriptor && _inputDescriptor.format && restrictToFormats !== undefined) {
                    // Take the instersection, as an argument has been supplied for restrictions
                    allowedFormats = Object.keys(_inputDescriptor.format).filter((k) => restrictToFormats.includes(k));
                }
                if (allowedFormats.includes(wvc.format)) {
                    // According to 18013-7 the docType MUST match the input descriptor ID
                    if (wvc.format === 'mso_mdoc') {
                        if (wvc.credential.docType !== _inputDescriptor.id) {
                            this.getResults().push(this.generateErrorResult(index, `$[${vcIndex}]`, wvc, PexMessages.INPUT_DESCRIPTOR_ID_MATCHES_MDOC_DOCTYPE_DIDNT_PASS));
                        }
                        if (_inputDescriptor.constraints?.fields?.some((field) => field.filter !== undefined)) {
                            this.getResults().push(this.generateErrorResult(index, `$[${vcIndex}]`, wvc, "Fields cannot have a 'filter' defined for mdoc credentials (ISO 18013-7)."));
                        }
                        if (_inputDescriptor.constraints?.fields?.some((field) => field.intent_to_retain === undefined)) {
                            this.getResults().push(this.generateErrorResult(index, `$[${vcIndex}]`, wvc, "Fields must have 'intent_to_retain' defined for mdoc credentials (ISO 18013-7)."));
                        }
                    }
                    this.getResults().push(this.generateSuccessResult(index, `$[${vcIndex}]`, wvc, `${wvc.format} is allowed from ${JSON.stringify(allowedFormats)}`));
                }
                else {
                    this.getResults().push(this.generateErrorResult(index, `$[${vcIndex}]`, wvc));
                }
            });
        });
        this.updatePresentationSubmission(pd);
    }
    generateErrorResult(idIdx, vcPath, wvc, message) {
        return {
            input_descriptor_path: `$.input_descriptors[${idIdx}]`,
            evaluator: this.getName(),
            status: Status.ERROR,
            message: message ?? PexMessages.FORMAT_RESTRICTION_DIDNT_PASS,
            verifiable_credential_path: vcPath,
            payload: {
                format: wvc.format,
            },
        };
    }
    generateSuccessResult(idIdx, vcPath, wvc, message) {
        return {
            input_descriptor_path: `$.input_descriptors[${idIdx}]`,
            evaluator: this.getName(),
            status: Status.INFO,
            message: message ?? PexMessages.FORMAT_RESTRICTION_PASSED,
            verifiable_credential_path: vcPath,
            payload: {
                format: wvc.format,
            },
        };
    }
}
