// @credo-ts/core/build/modules/kms/index.mjs


import { __exportAll } from "../../_virtual/rolldown_runtime.mjs";
import { KnownCoseSignatureAlgorithms } from "./cose/algorithm.mjs";
import "./cose/index.mjs";
import { KeyManagementError } from "./error/KeyManagementError.mjs";
import { KeyManagementAlgorithmNotSupportedError } from "./error/KeyManagementAlgorithmNotSupportedError.mjs";
import { KeyManagementKeyExistsError } from "./error/KeyManagementKeyExistsError.mjs";
import { KeyManagementKeyNotFoundError } from "./error/KeyManagementKeyNotFoundError.mjs";
import { getJwkHumanDescription } from "./jwk/humanDescription.mjs";
import { allowedEncryptionAlgsForKey, assertAllowedEncryptionAlgForKey, supportedEncryptionAlgsForKey } from "./jwk/alg/encryption.mjs";
import { allowedKeyDerivationAlgsForKey, assertAllowedKeyDerivationAlgForKey, supportedKeyDerivationAlgsForKey } from "./jwk/alg/keyDerivation.mjs";
import { allowedSigningAlgsForSigningKey, assertAllowedSigningAlgForKey, supportedSigningAlgsForKey } from "./jwk/alg/signing.mjs";
import { assertSupportedEncryptionAlgorithm, assertSupportedKeyAgreementAlgorithm } from "./jwk/assertSupported.mjs";
import { assertAsymmetricJwkKeyTypeMatches, assymetricJwkKeyTypeMatches, asymmetricPublicJwkMatches } from "./jwk/equals.mjs";
import { KnownJwaContentEncryptionAlgorithms, KnownJwaKeyAgreementAlgorithms, KnownJwaKeyEncryptionAlgorithms, KnownJwaSignatureAlgorithms } from "./jwk/jwa.mjs";
import { assertKeyAllowsDecrypt, assertKeyAllowsDerive, assertKeyAllowsEncrypt, assertKeyAllowsSign, assertKeyAllowsVerify, keyAllowsDecrypt, keyAllowsEncrypt, keyAllowsSign, keyAllowsVerify } from "./jwk/keyOps.mjs";
import { assertJwkAsymmetric, isJwkAsymmetric, publicJwkFromPrivateJwk } from "./jwk/knownJwk.mjs";
import { derEcSignatureToRaw, rawEcSignatureToDer } from "./jwk/kty/ec/ecSignature.mjs";
import { P256PublicJwk } from "./jwk/kty/ec/P256PublicJwk.mjs";
import { P384PublicJwk } from "./jwk/kty/ec/P384PublicJwk.mjs";
import { P521PublicJwk } from "./jwk/kty/ec/P521PublicJwk.mjs";
import { Secp256k1PublicJwk } from "./jwk/kty/ec/Secp256k1PublicJwk.mjs";
import { X25519PublicJwk } from "./jwk/kty/okp/X25519PublicJwk.mjs";
import { Ed25519PublicJwk } from "./jwk/kty/okp/Ed25519PublicJwk.mjs";
import { RsaPublicJwk } from "./jwk/kty/rsa/RsaPublicJwk.mjs";
import { legacyKeyIdFromPublicJwk } from "./legacy.mjs";
import { PublicJwk } from "./jwk/PublicJwk.mjs";
import "./jwk/index.mjs";
import { KeyManagementModuleConfig } from "./KeyManagementModuleConfig.mjs";
import { getKmsOperationHumanDescription } from "./options/KmsOperation.mjs";
import "./options/index.mjs";
import { KeyManagementApi } from "./KeyManagementApi.mjs";
import { KeyManagementModule } from "./KeyManagementModule.mjs";

//#region src/modules/kms/index.ts
var kms_exports = /* @__PURE__ */ __exportAll({
	Ed25519PublicJwk: () => Ed25519PublicJwk,
	KeyManagementAlgorithmNotSupportedError: () => KeyManagementAlgorithmNotSupportedError,
	KeyManagementApi: () => KeyManagementApi,
	KeyManagementError: () => KeyManagementError,
	KeyManagementKeyExistsError: () => KeyManagementKeyExistsError,
	KeyManagementKeyNotFoundError: () => KeyManagementKeyNotFoundError,
	KeyManagementModule: () => KeyManagementModule,
	KeyManagementModuleConfig: () => KeyManagementModuleConfig,
	KnownCoseSignatureAlgorithms: () => KnownCoseSignatureAlgorithms,
	KnownJwaContentEncryptionAlgorithms: () => KnownJwaContentEncryptionAlgorithms,
	KnownJwaKeyAgreementAlgorithms: () => KnownJwaKeyAgreementAlgorithms,
	KnownJwaKeyEncryptionAlgorithms: () => KnownJwaKeyEncryptionAlgorithms,
	KnownJwaSignatureAlgorithms: () => KnownJwaSignatureAlgorithms,
	P256PublicJwk: () => P256PublicJwk,
	P384PublicJwk: () => P384PublicJwk,
	P521PublicJwk: () => P521PublicJwk,
	PublicJwk: () => PublicJwk,
	RsaPublicJwk: () => RsaPublicJwk,
	Secp256k1PublicJwk: () => Secp256k1PublicJwk,
	X25519PublicJwk: () => X25519PublicJwk,
	allowedEncryptionAlgsForKey: () => allowedEncryptionAlgsForKey,
	allowedKeyDerivationAlgsForKey: () => allowedKeyDerivationAlgsForKey,
	allowedSigningAlgsForSigningKey: () => allowedSigningAlgsForSigningKey,
	assertAllowedEncryptionAlgForKey: () => assertAllowedEncryptionAlgForKey,
	assertAllowedKeyDerivationAlgForKey: () => assertAllowedKeyDerivationAlgForKey,
	assertAllowedSigningAlgForKey: () => assertAllowedSigningAlgForKey,
	assertAsymmetricJwkKeyTypeMatches: () => assertAsymmetricJwkKeyTypeMatches,
	assertJwkAsymmetric: () => assertJwkAsymmetric,
	assertKeyAllowsDecrypt: () => assertKeyAllowsDecrypt,
	assertKeyAllowsDerive: () => assertKeyAllowsDerive,
	assertKeyAllowsEncrypt: () => assertKeyAllowsEncrypt,
	assertKeyAllowsSign: () => assertKeyAllowsSign,
	assertKeyAllowsVerify: () => assertKeyAllowsVerify,
	assertSupportedEncryptionAlgorithm: () => assertSupportedEncryptionAlgorithm,
	assertSupportedKeyAgreementAlgorithm: () => assertSupportedKeyAgreementAlgorithm,
	assymetricJwkKeyTypeMatches: () => assymetricJwkKeyTypeMatches,
	asymmetricPublicJwkMatches: () => asymmetricPublicJwkMatches,
	derEcSignatureToRaw: () => derEcSignatureToRaw,
	getJwkHumanDescription: () => getJwkHumanDescription,
	getKmsOperationHumanDescription: () => getKmsOperationHumanDescription,
	isJwkAsymmetric: () => isJwkAsymmetric,
	keyAllowsDecrypt: () => keyAllowsDecrypt,
	keyAllowsEncrypt: () => keyAllowsEncrypt,
	keyAllowsSign: () => keyAllowsSign,
	keyAllowsVerify: () => keyAllowsVerify,
	legacyKeyIdFromPublicJwk: () => legacyKeyIdFromPublicJwk,
	publicJwkFromPrivateJwk: () => publicJwkFromPrivateJwk,
	rawEcSignatureToDer: () => rawEcSignatureToDer,
	supportedEncryptionAlgsForKey: () => supportedEncryptionAlgsForKey,
	supportedKeyDerivationAlgsForKey: () => supportedKeyDerivationAlgsForKey,
	supportedSigningAlgsForKey: () => supportedSigningAlgsForKey
});

//#endregion
export { Ed25519PublicJwk, KeyManagementAlgorithmNotSupportedError, KeyManagementApi, KeyManagementError, KeyManagementKeyExistsError, KeyManagementKeyNotFoundError, KeyManagementModule, KeyManagementModuleConfig, KnownCoseSignatureAlgorithms, KnownJwaContentEncryptionAlgorithms, KnownJwaKeyAgreementAlgorithms, KnownJwaKeyEncryptionAlgorithms, KnownJwaSignatureAlgorithms, P256PublicJwk, P384PublicJwk, P521PublicJwk, PublicJwk, RsaPublicJwk, Secp256k1PublicJwk, X25519PublicJwk, allowedEncryptionAlgsForKey, allowedKeyDerivationAlgsForKey, allowedSigningAlgsForSigningKey, assertAllowedEncryptionAlgForKey, assertAllowedKeyDerivationAlgForKey, assertAllowedSigningAlgForKey, assertAsymmetricJwkKeyTypeMatches, assertJwkAsymmetric, assertKeyAllowsDecrypt, assertKeyAllowsDerive, assertKeyAllowsEncrypt, assertKeyAllowsSign, assertKeyAllowsVerify, assertSupportedEncryptionAlgorithm, assertSupportedKeyAgreementAlgorithm, assymetricJwkKeyTypeMatches, asymmetricPublicJwkMatches, derEcSignatureToRaw, getJwkHumanDescription, getKmsOperationHumanDescription, isJwkAsymmetric, keyAllowsDecrypt, keyAllowsEncrypt, keyAllowsSign, keyAllowsVerify, kms_exports, legacyKeyIdFromPublicJwk, publicJwkFromPrivateJwk, rawEcSignatureToDer, supportedEncryptionAlgsForKey, supportedKeyDerivationAlgsForKey, supportedSigningAlgsForKey };
