// @credo-ts/core/build/modules/mdoc/repository/MdocRecord.mjs


import { JsonTransformer } from "../../../utils/JsonTransformer.mjs";
import "../../../utils/index.mjs";
import { BaseRecord } from "../../../storage/BaseRecord.mjs";
import { uuid } from "../../../utils/uuid.mjs";
import { CredentialMultiInstanceState } from "../../../utils/credentialUseTypes.mjs";
import { Mdoc } from "../Mdoc.mjs";

//#region src/modules/mdoc/repository/MdocRecord.ts
var MdocRecord = class MdocRecord extends BaseRecord {
	/**
	* Only here for class transformation. If base64Url is set we transform
	* it to the new mdocs array format
	*/
	set base64Url(base64Url) {
		this.credentialInstances = [{ issuerSignedBase64Url: base64Url }];
	}
	constructor(props) {
		super();
		this.type = MdocRecord.type;
		this.multiInstanceState = CredentialMultiInstanceState.SingleInstanceUsed;
		if (props) {
			this.id = props.id ?? uuid();
			this.createdAt = props.createdAt ?? /* @__PURE__ */ new Date();
			this.credentialInstances = props.credentialInstances;
			this.multiInstanceState = this.credentialInstances.length === 1 ? CredentialMultiInstanceState.SingleInstanceUnused : CredentialMultiInstanceState.MultiInstanceFirstUnused;
			this._tags = props.tags ?? {};
		}
	}
	get firstCredential() {
		const mdoc = Mdoc.fromBase64Url(this.credentialInstances[0].issuerSignedBase64Url);
		mdoc.deviceKeyId = this.credentialInstances[0].kmsKeyId ?? mdoc.deviceKey?.legacyKeyId;
		return mdoc;
	}
	static fromMdoc(mdoc) {
		return new MdocRecord({ credentialInstances: [{
			issuerSignedBase64Url: mdoc.base64Url,
			kmsKeyId: mdoc.deviceKeyId
		}] });
	}
	getTags() {
		const mdoc = this.firstCredential;
		const docType = mdoc.docType;
		const alg = mdoc.alg;
		return {
			...this._tags,
			docType,
			alg,
			multiInstanceState: this.multiInstanceState
		};
	}
	clone() {
		return JsonTransformer.fromJSON(JsonTransformer.toJSON(this), this.constructor);
	}
	/**
	* encoded is convenience method added to all credential records
	*/
	get encoded() {
		return this.credentialInstances[0].issuerSignedBase64Url;
	}
};
MdocRecord.type = "MdocRecord";

//#endregion
export { MdocRecord };
