// @animo-id/pex/dist/main/lib/evaluation/handlers/formatRestrictionEvaluationHandler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormatRestrictionEvaluationHandler = void 0;
const ConstraintUtils_1 = require("../../ConstraintUtils");
const Messages_1 = __importDefault(require("../../types/Messages"));
const abstractEvaluationHandler_1 = require("./abstractEvaluationHandler");
class FormatRestrictionEvaluationHandler extends abstractEvaluationHandler_1.AbstractEvaluationHandler {
    constructor(client) {
        super(client);
    }
    getName() {
        return 'FormatRestrictionEvaluation';
    }
    handle(pd, wrappedVcs) {
        const restrictToFormats = this.client.restrictToFormats ? Object.keys(this.client.restrictToFormats) : undefined;
        pd.input_descriptors.forEach((_inputDescriptor, index) => {
            wrappedVcs.forEach((wvc, vcIndex) => {
                var _a, _b, _c, _d;
                const formats = 'format' in _inputDescriptor && _inputDescriptor.format ? Object.keys(_inputDescriptor.format) : [wvc.format];
                let allowedFormats = restrictToFormats !== null && restrictToFormats !== void 0 ? restrictToFormats : formats;
                if ('format' in _inputDescriptor && _inputDescriptor.format && restrictToFormats !== undefined) {
                    // Take the instersection, as an argument has been supplied for restrictions
                    allowedFormats = Object.keys(_inputDescriptor.format).filter((k) => restrictToFormats.includes(k));
                }
                if (allowedFormats.includes(wvc.format)) {
                    // According to 18013-7 the docType MUST match the input descriptor ID
                    if (wvc.format === 'mso_mdoc') {
                        if (wvc.credential.docType !== _inputDescriptor.id) {
                            this.getResults().push(this.generateErrorResult(index, `$[${vcIndex}]`, wvc, Messages_1.default.INPUT_DESCRIPTOR_ID_MATCHES_MDOC_DOCTYPE_DIDNT_PASS));
                        }
                        if ((_b = (_a = _inputDescriptor.constraints) === null || _a === void 0 ? void 0 : _a.fields) === null || _b === void 0 ? void 0 : _b.some((field) => field.filter !== undefined)) {
                            this.getResults().push(this.generateErrorResult(index, `$[${vcIndex}]`, wvc, "Fields cannot have a 'filter' defined for mdoc credentials (ISO 18013-7)."));
                        }
                        if ((_d = (_c = _inputDescriptor.constraints) === null || _c === void 0 ? void 0 : _c.fields) === null || _d === void 0 ? void 0 : _d.some((field) => field.intent_to_retain === undefined)) {
                            this.getResults().push(this.generateErrorResult(index, `$[${vcIndex}]`, wvc, "Fields must have 'intent_to_retain' defined for mdoc credentials (ISO 18013-7)."));
                        }
                    }
                    this.getResults().push(this.generateSuccessResult(index, `$[${vcIndex}]`, wvc, `${wvc.format} is allowed from ${JSON.stringify(allowedFormats)}`));
                }
                else {
                    this.getResults().push(this.generateErrorResult(index, `$[${vcIndex}]`, wvc));
                }
            });
        });
        this.updatePresentationSubmission(pd);
    }
    generateErrorResult(idIdx, vcPath, wvc, message) {
        return {
            input_descriptor_path: `$.input_descriptors[${idIdx}]`,
            evaluator: this.getName(),
            status: ConstraintUtils_1.Status.ERROR,
            message: message !== null && message !== void 0 ? message : Messages_1.default.FORMAT_RESTRICTION_DIDNT_PASS,
            verifiable_credential_path: vcPath,
            payload: {
                format: wvc.format,
            },
        };
    }
    generateSuccessResult(idIdx, vcPath, wvc, message) {
        return {
            input_descriptor_path: `$.input_descriptors[${idIdx}]`,
            evaluator: this.getName(),
            status: ConstraintUtils_1.Status.INFO,
            message: message !== null && message !== void 0 ? message : Messages_1.default.FORMAT_RESTRICTION_PASSED,
            verifiable_credential_path: vcPath,
            payload: {
                format: wvc.format,
            },
        };
    }
}
exports.FormatRestrictionEvaluationHandler = FormatRestrictionEvaluationHandler;
