// @animo-id/pex/dist/main/lib/evaluation/handlers/uriEvaluationHandler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UriEvaluationHandler = void 0;
const jsonpath_1 = require("@astronautlabs/jsonpath");
const ssi_types_1 = require("@sphereon/ssi-types");
const nanoid_1 = require("nanoid");
const ConstraintUtils_1 = require("../../ConstraintUtils");
const types_1 = require("../../types");
const Messages_1 = __importDefault(require("../../types/Messages"));
const PexCredentialMapper_1 = require("../../types/PexCredentialMapper");
const abstractEvaluationHandler_1 = require("./abstractEvaluationHandler");
class UriEvaluationHandler extends abstractEvaluationHandler_1.AbstractEvaluationHandler {
    constructor(client) {
        super(client);
    }
    getName() {
        return 'UriEvaluation';
    }
    handle(definition, wrappedVcs) {
        var _a, _b, _c;
        // This filter is removed in V2
        definition.input_descriptors.forEach((inDesc, descriptorIdx) => {
            const uris = definition.getVersion() !== types_1.PEVersion.v2 ? inDesc.schema.map((so) => so.uri) : [];
            wrappedVcs.forEach((wvc, wrappedVCIdx) => {
                const vcUris = UriEvaluationHandler.buildVcContextAndSchemaUris(wvc.credential, definition.getVersion());
                this.evaluateUris(wvc, vcUris, uris, descriptorIdx, wrappedVCIdx, definition.getVersion());
            });
        });
        const definitionAllowsDataIntegrity = ((_a = definition.format) === null || _a === void 0 ? void 0 : _a.di) || ((_b = definition.format) === null || _b === void 0 ? void 0 : _b.di_vc) || ((_c = definition.format) === null || _c === void 0 ? void 0 : _c.di_vp);
        const descriptorMap = this.getResults()
            .filter((result) => result.status === ConstraintUtils_1.Status.INFO)
            .map((result) => {
            var _a, _b, _c, _d;
            let format = (_a = result.payload) === null || _a === void 0 ? void 0 : _a.format;
            // This checks if the new data integrity format should be used.
            // That may be the case if the input descriptor points to credentials that are in ldp_vc or ldp format,
            // and the presentation definition allows data integrity.
            if (definitionAllowsDataIntegrity && (format === 'ldp_vc' || format === 'ldp')) {
                const wvcs = jsonpath_1.JSONPath.nodes(wrappedVcs, result.verifiable_credential_path).map((node) => node.value);
                // check if all vc's have a data integrity proof
                const vcDataIntegrityProofs = wvcs.map((vc) => {
                    if (vc.type !== ssi_types_1.OriginalType.JSONLD || !vc.credential.proof)
                        return [];
                    const proofs = Array.isArray(vc.credential.proof) ? vc.credential.proof : [vc.credential.proof];
                    const dataIntegrityProofs = proofs.filter((proof) => proof.type === 'DataIntegrityProof' && proof.cryptosuite !== undefined);
                    return dataIntegrityProofs;
                });
                // determine the common cryptosuites of all vc's
                const commonCryptosuites = vcDataIntegrityProofs.reduce((a, b) => a.filter((c) => b.includes(c)));
                // the input descriptor should also allow data integrity
                const inputDescriptor = jsonpath_1.JSONPath.nodes(definition, result.input_descriptor_path)[0].value;
                const inputDescriptorAllowsDataIntegrity = !inputDescriptor['format'] || ((_b = inputDescriptor === null || inputDescriptor === void 0 ? void 0 : inputDescriptor.format) === null || _b === void 0 ? void 0 : _b.di) || ((_c = inputDescriptor === null || inputDescriptor === void 0 ? void 0 : inputDescriptor.format) === null || _c === void 0 ? void 0 : _c.di_vc) || ((_d = inputDescriptor === null || inputDescriptor === void 0 ? void 0 : inputDescriptor.format) === null || _d === void 0 ? void 0 : _d.di_vp);
                if (commonCryptosuites.length > 0 && inputDescriptorAllowsDataIntegrity) {
                    format = 'di_vc';
                }
            }
            const inputDescriptor = jsonpath_1.JSONPath.nodes(definition, result.input_descriptor_path)[0].value;
            return {
                id: inputDescriptor.id,
                format,
                path: result.verifiable_credential_path,
            };
        });
        // The presentation submission is being created in this handler, then updated in subsequent handler.
        // TODO: This approach needs to be refactored for a new Major version.
        // Also there is no apparent need for the indirection and state in this class.
        // Simply do the first loops and amend the presentation submission in every loop.
        if (this.client.generatePresentationSubmission && (!this.presentationSubmission || Object.keys(this.presentationSubmission).length === 0)) {
            this.presentationSubmission = {
                id: (0, nanoid_1.nanoid)(),
                definition_id: definition.id,
                descriptor_map: descriptorMap,
            };
        }
    }
    evaluateUris(wvc, verifiableCredentialUris, inputDescriptorsUris, idIdx, vcIdx, pdVersion) {
        let hasAnyMatch = false;
        if (pdVersion === types_1.PEVersion.v1) {
            for (let i = 0; i < inputDescriptorsUris.length; i++) {
                if (UriEvaluationHandler.containsHashlink(inputDescriptorsUris[i])) {
                    this.getResults().push(this.createWarnResultObject(idIdx, vcIdx));
                }
            }
            for (let i = 0; i < verifiableCredentialUris.length; i++) {
                if (inputDescriptorsUris.find((el) => el === verifiableCredentialUris[i]) != undefined) {
                    hasAnyMatch = true;
                }
            }
        }
        else {
            hasAnyMatch = true;
        }
        if (hasAnyMatch) {
            this.getResults().push(this.createSuccessResultObject(wvc, inputDescriptorsUris, idIdx, vcIdx));
        }
        else {
            this.getResults().push(this.createErrorResultObject(wvc, inputDescriptorsUris, idIdx, vcIdx));
        }
    }
    static buildVcContextAndSchemaUris(credential, version) {
        const uris = [];
        // W3C credential
        if (PexCredentialMapper_1.PexCredentialMapper.isW3cCredential(credential)) {
            if (Array.isArray(credential['@context'])) {
                credential['@context'].forEach((value) => uris.push(value));
            }
            else {
                uris.push(credential['@context']);
            }
            if (Array.isArray(credential.credentialSchema) && credential.credentialSchema.length > 0) {
                credential.credentialSchema.forEach((element) => uris.push(element.id));
            }
            else if (credential.credentialSchema) {
                uris.push(credential.credentialSchema.id);
            }
            if (version === types_1.PEVersion.v1) {
                // JWT VC Profile and MS Entry Verified ID do use the schema from V1 to match against types in the VC
                Array.isArray(credential.type)
                    ? credential.type.forEach((type) => uris.push(type))
                    : credential.type
                        ? uris.push(credential.type)
                        : undefined;
            }
        }
        // NOTE: we add the `vct` field of an SD-JWT to the list of uris, to allow SD-JWT
        // to work with PEX v1 in the same way that JWT vcs can work with pex v1. If we don't
        // add this, then SD-JWTs can only be used with PEX v2.
        if (PexCredentialMapper_1.PexCredentialMapper.isSdJwtDecodedCredential(credential)) {
            if (version === types_1.PEVersion.v1) {
                uris.push(credential.decodedPayload.vct);
            }
        }
        return uris;
    }
    createSuccessResultObject(wvc, inputDescriptorsUris, idIdx, vcIdx) {
        const result = this.createResult(idIdx, vcIdx);
        result.status = ConstraintUtils_1.Status.INFO;
        result.message = Messages_1.default.URI_EVALUATION_PASSED;
        result.payload = {
            format: wvc.format,
            vcContext: PexCredentialMapper_1.PexCredentialMapper.isW3cCredential(wvc.credential) ? wvc.credential['@context'] : undefined,
            vcCredentialSchema: PexCredentialMapper_1.PexCredentialMapper.isW3cCredential(wvc.credential) ? wvc.credential.credentialSchema : undefined,
            inputDescriptorsUris,
        };
        return result;
    }
    createErrorResultObject(wvc, inputDescriptorsUris, idIdx, vcIdx) {
        const result = this.createResult(idIdx, vcIdx);
        result.status = ConstraintUtils_1.Status.ERROR;
        result.message = Messages_1.default.URI_EVALUATION_DIDNT_PASS;
        result.payload = {
            format: wvc.format,
            vcContext: PexCredentialMapper_1.PexCredentialMapper.isW3cCredential(wvc.credential) ? wvc.credential['@context'] : undefined,
            vcCredentialSchema: PexCredentialMapper_1.PexCredentialMapper.isW3cCredential(wvc.credential) ? wvc.credential.credentialSchema : undefined,
            inputDescriptorsUris,
        };
        return result;
    }
    createWarnResultObject(idIdx, vcIdx) {
        const result = this.createResult(idIdx, vcIdx);
        result.status = ConstraintUtils_1.Status.WARN;
        result.message = Messages_1.default.URI_EVALUATION_DIDNT_PASS;
        result.payload = Messages_1.default.INPUT_DESCRIPTOR_CONTEXT_CONTAINS_HASHLINK_VERIFICATION_NOT_SUPPORTED;
        return result;
    }
    createResult(idIdx, vcIdx) {
        return {
            input_descriptor_path: `$.input_descriptors[${idIdx}]`,
            verifiable_credential_path: `$[${vcIdx}]`,
            evaluator: this.getName(),
            status: ConstraintUtils_1.Status.INFO,
            message: undefined,
        };
    }
    static containsHashlink(url) {
        return !(url.matchAll(UriEvaluationHandler.HASHLINK_QUERY_URL_REGEX).next().done &&
            url.matchAll(UriEvaluationHandler.HASHLINK_URL_ENCODED_REGEX).next().done);
    }
}
exports.UriEvaluationHandler = UriEvaluationHandler;
UriEvaluationHandler.HASHLINK_URL_ENCODED_REGEX = /hl:[a-zA-Z0-9]+:[a-zA-Z0-9]+/g;
UriEvaluationHandler.HASHLINK_QUERY_URL_REGEX = /https*?:\/\/(www\.)?[-a-zA-Z0-9@:%._+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_+.~#?&/=]*)(hl=[a-zA-Z0-9]+)/g;
