// @credo-ts/core/build/modules/kms/jwk/kty/ec/P521PublicJwk.mjs


import { KnownJwaKeyAgreementAlgorithms, KnownJwaSignatureAlgorithms } from "../../jwa.mjs";
import { ecPublicJwkToPublicKey, ecPublicKeyToPublicJwk } from "./ecPublicKey.mjs";

//#region src/modules/kms/jwk/kty/ec/P521PublicJwk.ts
var P521PublicJwk = class P521PublicJwk {
	constructor(jwk) {
		this.jwk = jwk;
		this.supportedEncryptionKeyAgreementAlgorithms = P521PublicJwk.supportedEncryptionKeyAgreementAlgorithms;
		this.supportedSignatureAlgorithms = P521PublicJwk.supportedSignatureAlgorithms;
		this.multicodecPrefix = P521PublicJwk.multicodecPrefix;
	}
	get publicKey() {
		return {
			crv: this.jwk.crv,
			kty: this.jwk.kty,
			publicKey: ecPublicJwkToPublicKey(this.jwk)
		};
	}
	get compressedPublicKey() {
		return {
			crv: this.jwk.crv,
			kty: this.jwk.kty,
			publicKey: ecPublicJwkToPublicKey(this.jwk, { compressed: true })
		};
	}
	get multicodec() {
		return ecPublicJwkToPublicKey(this.jwk, { compressed: true });
	}
	static fromPublicKey(publicKey) {
		return new P521PublicJwk(ecPublicKeyToPublicJwk(publicKey, "P-521"));
	}
	static fromMulticodec(multicodec) {
		return new P521PublicJwk(ecPublicKeyToPublicJwk(multicodec, "P-521"));
	}
};
P521PublicJwk.supportedSignatureAlgorithms = [KnownJwaSignatureAlgorithms.ES512];
P521PublicJwk.supportedEncryptionKeyAgreementAlgorithms = [KnownJwaKeyAgreementAlgorithms.ECDH_ES];
P521PublicJwk.multicodecPrefix = 4610;

//#endregion
export { P521PublicJwk };
