// @credo-ts/node/build/kms/crypto/createKey.mjs
import { Kms } from "@credo-ts/core";
import { generateKeyPair, randomBytes } from "node:crypto";
import { promisify } from "node:util";

//#region src/kms/crypto/createKey.ts
const generateKeyPair$1 = promisify(generateKeyPair);
const nodeSupportedEcCrvs = [
	"P-256",
	"P-384",
	"P-521",
	"secp256k1"
];
function assertNodeSupportedEcCrv(options) {
	if (!nodeSupportedEcCrvs.includes(options.crv)) throw new Kms.KeyManagementAlgorithmNotSupportedError(`crv '${options.crv}' for kty '${options.kty}'`, "node");
}
async function createEcKey({ crv }) {
	const { publicKey, privateKey } = await generateKeyPair$1("ec", { namedCurve: crv });
	return {
		privateJwk: privateKey.export({ format: "jwk" }),
		publicJwk: publicKey.export({ format: "jwk" })
	};
}
async function createRsaKey({ modulusLength }) {
	const { publicKey, privateKey } = await generateKeyPair$1("rsa", { modulusLength });
	return {
		privateJwk: privateKey.export({ format: "jwk" }),
		publicJwk: publicKey.export({ format: "jwk" })
	};
}
const nodeSupportedOkpCrvs = ["Ed25519", "X25519"];
function assertNodeSupportedOkpCrv(options) {
	if (!nodeSupportedOkpCrvs.includes(options.crv)) throw new Kms.KeyManagementAlgorithmNotSupportedError(`crv '${options.crv}' for kty '${options.kty}'`, "node");
}
async function createOkpKey({ crv }) {
	const { publicKey, privateKey } = crv === "Ed25519" ? await generateKeyPair$1("ed25519") : await generateKeyPair$1("x25519");
	return {
		privateJwk: privateKey.export({ format: "jwk" }),
		publicJwk: publicKey.export({ format: "jwk" })
	};
}
const nodeSupportedOctAlgorithms = ["aes", "hmac"];
function assertNodeSupportedOctAlgorithm(options) {
	if (!nodeSupportedOctAlgorithms.includes(options.algorithm)) throw new Kms.KeyManagementAlgorithmNotSupportedError(`algorithm '${options.algorithm}' for kty '${options.kty}'`, "node");
}
async function createOctKey(options) {
	const privateJwk = {
		kty: "oct",
		k: randomBytes(options.length >> 3).toString("base64url")
	};
	const { k, ...publicJwk } = privateJwk;
	return {
		privateJwk,
		publicJwk
	};
}

//#endregion
export { assertNodeSupportedEcCrv, assertNodeSupportedOctAlgorithm, assertNodeSupportedOkpCrv, createEcKey, createOctKey, createOkpKey, createRsaKey };
