// @credo-ts/didcomm/build/modules/message-pickup/DidCommMessagePickupModule.mjs
import { DidCommFeatureRegistry } from "../../DidCommFeatureRegistry.mjs";
import { DidCommMessageHandlerRegistry } from "../../DidCommMessageHandlerRegistry.mjs";
import { DidCommMessagePickupModuleConfig } from "./DidCommMessagePickupModuleConfig.mjs";
import { DidCommMessagePickupSessionService } from "./services/DidCommMessagePickupSessionService.mjs";
import { DidCommMessagePickupApi } from "./DidCommMessagePickupApi.mjs";
import { DidCommMessagePickupV1Protocol } from "./protocol/v1/DidCommMessagePickupV1Protocol.mjs";
import "./services/index.mjs";
import { DidCommMessagePickupV2Protocol } from "./protocol/v2/DidCommMessagePickupV2Protocol.mjs";
import "./protocol/index.mjs";

//#region src/modules/message-pickup/DidCommMessagePickupModule.ts
var DidCommMessagePickupModule = class {
	constructor(config) {
		this.api = DidCommMessagePickupApi;
		this.config = new DidCommMessagePickupModuleConfig({
			...config,
			protocols: config?.protocols ?? [new DidCommMessagePickupV1Protocol(), new DidCommMessagePickupV2Protocol()]
		});
	}
	/**
	* Registers the dependencies of the message pickup answer module on the dependency manager.
	*/
	register(dependencyManager) {
		dependencyManager.registerInstance(DidCommMessagePickupModuleConfig, this.config);
		dependencyManager.registerSingleton(DidCommMessagePickupSessionService);
	}
	async initialize(agentContext) {
		const messageHandlerRegistry = agentContext.dependencyManager.resolve(DidCommMessageHandlerRegistry);
		const featureRegistry = agentContext.dependencyManager.resolve(DidCommFeatureRegistry);
		for (const protocol of this.config.protocols) protocol.register(messageHandlerRegistry, featureRegistry);
	}
	async onInitializeContext(agentContext) {
		if (!agentContext.isRootAgentContext) return;
		agentContext.dependencyManager.resolve(DidCommMessagePickupSessionService).start(agentContext);
	}
};

//#endregion
export { DidCommMessagePickupModule };
