// @credo-ts/didcomm/build/modules/message-pickup/protocol/v2/messages/DidCommDeliveryRequestV2Message.mjs
import { __decorate } from "../../../../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorate.mjs";
import { IsValidMessageType, parseMessageType } from "../../../../../util/messageType.mjs";
import { __decorateMetadata } from "../../../../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorateMetadata.mjs";
import { ReturnRouteTypes } from "../../../../../decorators/transport/TransportDecorator.mjs";
import { DidCommMessage } from "../../../../../DidCommMessage.mjs";
import { IsInt, IsOptional, IsString } from "class-validator";
import { Expose } from "class-transformer";

//#region src/modules/message-pickup/protocol/v2/messages/DidCommDeliveryRequestV2Message.ts
var DidCommDeliveryRequestV2Message = class DidCommDeliveryRequestV2Message extends DidCommMessage {
	constructor(options) {
		super();
		this.allowQueueTransport = false;
		this.type = DidCommDeliveryRequestV2Message.type.messageTypeUri;
		if (options) {
			this.id = options.id || this.generateId();
			this.recipientKey = options.recipientKey;
			this.limit = options.limit;
		}
		this.setReturnRouting(ReturnRouteTypes.all);
	}
};
DidCommDeliveryRequestV2Message.type = parseMessageType("https://didcomm.org/messagepickup/2.0/delivery-request");
__decorate([IsValidMessageType(DidCommDeliveryRequestV2Message.type), __decorateMetadata("design:type", Object)], DidCommDeliveryRequestV2Message.prototype, "type", void 0);
__decorate([
	IsString(),
	IsOptional(),
	Expose({ name: "recipient_key" }),
	__decorateMetadata("design:type", String)
], DidCommDeliveryRequestV2Message.prototype, "recipientKey", void 0);
__decorate([IsInt(), __decorateMetadata("design:type", Number)], DidCommDeliveryRequestV2Message.prototype, "limit", void 0);

//#endregion
export { DidCommDeliveryRequestV2Message };
