// @credo-ts/core/build/modules/x509/X509Api.mjs


import { AgentContext } from "../../agent/context/AgentContext.mjs";
import { injectable } from "../../plugins/index.mjs";
import { __decorateMetadata } from "../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorateMetadata.mjs";
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorate.mjs";
import "../../agent/index.mjs";
import { X509ModuleConfig } from "./X509ModuleConfig.mjs";
import { X509Service } from "./X509Service.mjs";

//#region src/modules/x509/X509Api.ts
var _ref, _ref2;
let X509Api = class X509Api {
	constructor(agentContext, config) {
		this.agentContext = agentContext;
		this.config = config;
	}
	/**
	* Creates a X.509 certificate.
	*
	* @param options X509CreateCertificateOptions
	*/
	async createCertificate(options) {
		return await X509Service.createCertificate(this.agentContext, options);
	}
	async createCertificateSigningRequest(options) {
		return await X509Service.createCertificateSigningRequest(this.agentContext, options);
	}
	parseCertificateSigningRequest(options) {
		return X509Service.parseCertificateSigningRequest(options);
	}
	/**
	* Validate a certificate chain.
	*
	* @param options X509ValidateCertificateChainOptions
	*/
	async validateCertificateChain(options) {
		return await X509Service.validateCertificateChain(this.agentContext, options);
	}
};
X509Api = __decorate([injectable(), __decorateMetadata("design:paramtypes", [typeof (_ref = typeof AgentContext !== "undefined" && AgentContext) === "function" ? _ref : Object, typeof (_ref2 = typeof X509ModuleConfig !== "undefined" && X509ModuleConfig) === "function" ? _ref2 : Object])], X509Api);

//#endregion
export { X509Api };
