// @credo-ts/didcomm/build/decorators/ack/AckDecoratorExtension.mjs
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorate.mjs";
import { __decorateMetadata } from "../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorateMetadata.mjs";
import { AckDecorator, AckValues } from "./AckDecorator.mjs";
import { IsInstance, IsOptional, ValidateNested } from "class-validator";
import { Expose, Type } from "class-transformer";

//#region src/decorators/ack/AckDecoratorExtension.ts
function AckDecorated(Base) {
	var _ref;
	class AckDecoratorExtension extends Base {
		setPleaseAck(on = [AckValues.Receipt]) {
			this.pleaseAck = new AckDecorator({ on });
		}
		getPleaseAck() {
			return this.pleaseAck;
		}
		requiresAck() {
			return this.pleaseAck !== void 0;
		}
	}
	__decorate([
		Expose({ name: "~please_ack" }),
		Type(() => AckDecorator),
		ValidateNested(),
		IsInstance(AckDecorator),
		IsOptional(),
		__decorateMetadata("design:type", typeof (_ref = typeof AckDecorator !== "undefined" && AckDecorator) === "function" ? _ref : Object)
	], AckDecoratorExtension.prototype, "pleaseAck", void 0);
	return AckDecoratorExtension;
}

//#endregion
export { AckDecorated };
