// @credo-ts/core/build/crypto/webcrypto/algorithmIdentifiers.mjs


import { id_ecPublicKey, id_secp256r1, id_secp384r1, id_secp521r1 } from "@peculiar/asn1-ecc";
import { AsnObjectIdentifierConverter } from "@peculiar/asn1-schema";
import { AlgorithmIdentifier } from "@peculiar/asn1-x509";
import { id_rsaEncryption } from "@peculiar/asn1-rsa";

//#region src/crypto/webcrypto/algorithmIdentifiers.ts
const ecPublicKeyAlgorithmIdentifier = (objectId) => new AlgorithmIdentifier({
	algorithm: id_ecPublicKey,
	parameters: AsnObjectIdentifierConverter.toASN(objectId).toBER()
});
/**
*
* https://oid-rep.orange-labs.fr/get/1.2.840.10045.3.1.7
*
*/
const ecPublicKeyWithP256AlgorithmIdentifier = ecPublicKeyAlgorithmIdentifier(id_secp256r1);
/**
*
* https://oid-rep.orange-labs.fr/get/1.3.132.0.34
*
*/
const ecPublicKeyWithP384AlgorithmIdentifier = ecPublicKeyAlgorithmIdentifier(id_secp384r1);
/**
*
* https://oid-rep.orange-labs.fr/get/1.3.132.0.35
*
*/
const ecPublicKeyWithP521AlgorithmIdentifier = ecPublicKeyAlgorithmIdentifier(id_secp521r1);
/**
*
* https://oid-rep.orange-labs.fr/get/1.3.132.0.10
*
*/
const ecPublicKeyWithK256AlgorithmIdentifier = ecPublicKeyAlgorithmIdentifier("1.3.132.0.10");
/**
*
* from: https://datatracker.ietf.org/doc/html/rfc8410#section-3
*
*/
const ed25519AlgorithmIdentifier = new AlgorithmIdentifier({ algorithm: "1.3.101.112" });
/**
*
* from: https://datatracker.ietf.org/doc/html/rfc8410#section-3
*
*/
const x25519AlgorithmIdentifier = new AlgorithmIdentifier({ algorithm: "1.3.101.110" });
/**
*
* RSA algorithm identifier
* https://oid-rep.orange-labs.fr/get/1.2.840.113549.1.1.1
*
*/
const rsaKeyAlgorithmIdentifier = new AlgorithmIdentifier({
	algorithm: id_rsaEncryption,
	parameters: null
});

//#endregion
export { ecPublicKeyWithK256AlgorithmIdentifier, ecPublicKeyWithP256AlgorithmIdentifier, ecPublicKeyWithP384AlgorithmIdentifier, ecPublicKeyWithP521AlgorithmIdentifier, ed25519AlgorithmIdentifier, rsaKeyAlgorithmIdentifier, x25519AlgorithmIdentifier };
