// @credo-ts/core/build/modules/mdoc/MdocService.mjs


import { __decorateMetadata } from "../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorateMetadata.mjs";
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorate.mjs";
import "../../agent/index.mjs";
import { Mdoc } from "./Mdoc.mjs";
import { MdocRepository } from "./repository/MdocRepository.mjs";
import "./repository/index.mjs";
import { MdocDeviceResponse } from "./MdocDeviceResponse.mjs";
import { injectable } from "tsyringe";

//#region src/modules/mdoc/MdocService.ts
var _ref;
let MdocService = class MdocService {
	constructor(mdocRepository) {
		this.MdocRepository = mdocRepository;
	}
	mdocFromBase64Url(hexEncodedMdoc) {
		return Mdoc.fromBase64Url(hexEncodedMdoc);
	}
	signMdoc(agentContext, options) {
		return Mdoc.sign(agentContext, options);
	}
	async verifyMdoc(agentContext, mdoc, options) {
		return await mdoc.verify(agentContext, options);
	}
	async createDeviceResponse(agentContext, options) {
		return MdocDeviceResponse.createDeviceResponse(agentContext, options);
	}
	async createPresentationDefinitionDeviceResponse(agentContext, options) {
		return MdocDeviceResponse.createPresentationDefinitionDeviceResponse(agentContext, options);
	}
	async verifyDeviceResponse(agentContext, options) {
		return MdocDeviceResponse.fromBase64Url(options.deviceResponse).verify(agentContext, options);
	}
	async store(agentContext, options) {
		await this.MdocRepository.save(agentContext, options.record);
		return options.record;
	}
	async getById(agentContext, id) {
		return await this.MdocRepository.getById(agentContext, id);
	}
	async getAll(agentContext) {
		return await this.MdocRepository.getAll(agentContext);
	}
	async findByQuery(agentContext, query, queryOptions) {
		return await this.MdocRepository.findByQuery(agentContext, query, queryOptions);
	}
	async deleteById(agentContext, id) {
		await this.MdocRepository.deleteById(agentContext, id);
	}
	async update(agentContext, mdocRecord) {
		await this.MdocRepository.update(agentContext, mdocRecord);
	}
};
MdocService = __decorate([injectable(), __decorateMetadata("design:paramtypes", [typeof (_ref = typeof MdocRepository !== "undefined" && MdocRepository) === "function" ? _ref : Object])], MdocService);

//#endregion
export { MdocService };
