// @credo-ts/didcomm/build/modules/credentials/protocol/revocation-notification/messages/DidCommRevocationNotificationV1Message.mjs
import { __decorate } from "../../../../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorate.mjs";
import { IsValidMessageType, parseMessageType } from "../../../../../util/messageType.mjs";
import { __decorateMetadata } from "../../../../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorateMetadata.mjs";
import "../../../../../decorators/ack/AckDecorator.mjs";
import { DidCommMessage } from "../../../../../DidCommMessage.mjs";
import { IsOptional, IsString } from "class-validator";
import { Expose } from "class-transformer";

//#region src/modules/credentials/protocol/revocation-notification/messages/DidCommRevocationNotificationV1Message.ts
var DidCommRevocationNotificationV1Message = class DidCommRevocationNotificationV1Message extends DidCommMessage {
	constructor(options) {
		super();
		this.type = DidCommRevocationNotificationV1Message.type.messageTypeUri;
		if (options) {
			this.issueThread = options.issueThread;
			this.id = options.id ?? this.generateId();
			this.comment = options.comment;
			this.pleaseAck = options.pleaseAck;
		}
	}
};
DidCommRevocationNotificationV1Message.type = parseMessageType("https://didcomm.org/revocation_notification/1.0/revoke");
__decorate([IsValidMessageType(DidCommRevocationNotificationV1Message.type), __decorateMetadata("design:type", Object)], DidCommRevocationNotificationV1Message.prototype, "type", void 0);
__decorate([
	IsString(),
	IsOptional(),
	__decorateMetadata("design:type", String)
], DidCommRevocationNotificationV1Message.prototype, "comment", void 0);
__decorate([
	Expose({ name: "thread_id" }),
	IsString(),
	__decorateMetadata("design:type", String)
], DidCommRevocationNotificationV1Message.prototype, "issueThread", void 0);

//#endregion
export { DidCommRevocationNotificationV1Message };
