// @credo-ts/didcomm/build/modules/credentials/protocol/revocation-notification/messages/DidCommRevocationNotificationV2Message.mjs
import { __decorate } from "../../../../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorate.mjs";
import { IsValidMessageType, parseMessageType } from "../../../../../util/messageType.mjs";
import { __decorateMetadata } from "../../../../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorateMetadata.mjs";
import "../../../../../decorators/ack/AckDecorator.mjs";
import { DidCommMessage } from "../../../../../DidCommMessage.mjs";
import { IsOptional, IsString } from "class-validator";
import { Expose } from "class-transformer";

//#region src/modules/credentials/protocol/revocation-notification/messages/DidCommRevocationNotificationV2Message.ts
var DidCommRevocationNotificationV2Message = class DidCommRevocationNotificationV2Message extends DidCommMessage {
	constructor(options) {
		super();
		this.type = DidCommRevocationNotificationV2Message.type.messageTypeUri;
		if (options) {
			this.revocationFormat = options.revocationFormat;
			this.credentialId = options.credentialId;
			this.id = options.id ?? this.generateId();
			this.comment = options.comment;
			this.pleaseAck = options.pleaseAck;
		}
	}
};
DidCommRevocationNotificationV2Message.type = parseMessageType("https://didcomm.org/revocation_notification/2.0/revoke");
__decorate([IsValidMessageType(DidCommRevocationNotificationV2Message.type), __decorateMetadata("design:type", Object)], DidCommRevocationNotificationV2Message.prototype, "type", void 0);
__decorate([
	IsString(),
	IsOptional(),
	__decorateMetadata("design:type", String)
], DidCommRevocationNotificationV2Message.prototype, "comment", void 0);
__decorate([
	Expose({ name: "revocation_format" }),
	IsString(),
	__decorateMetadata("design:type", String)
], DidCommRevocationNotificationV2Message.prototype, "revocationFormat", void 0);
__decorate([
	Expose({ name: "credential_id" }),
	IsString(),
	__decorateMetadata("design:type", String)
], DidCommRevocationNotificationV2Message.prototype, "credentialId", void 0);

//#endregion
export { DidCommRevocationNotificationV2Message };
