// @credo-ts/didcomm/build/modules/oob/DidCommOutOfBandModule.mjs
import { DidCommFeatureRegistry } from "../../DidCommFeatureRegistry.mjs";
import { DidCommMessageHandlerRegistry } from "../../DidCommMessageHandlerRegistry.mjs";
import { DidCommOutOfBandRepository } from "./repository/DidCommOutOfBandRepository.mjs";
import { DidCommProtocol } from "../../models/features/DidCommProtocol.mjs";
import "../../models/index.mjs";
import "./repository/index.mjs";
import { DidCommOutOfBandService } from "./DidCommOutOfBandService.mjs";
import { DidCommOutOfBandApi } from "./DidCommOutOfBandApi.mjs";
import { DidCommHandshakeReuseHandler } from "./handlers/DidCommHandshakeReuseHandler.mjs";
import "./handlers/index.mjs";
import { DidCommHandshakeReuseAcceptedHandler } from "./handlers/DidCommHandshakeReuseAcceptedHandler.mjs";

//#region src/modules/oob/DidCommOutOfBandModule.ts
var DidCommOutOfBandModule = class {
	constructor() {
		this.api = DidCommOutOfBandApi;
	}
	/**
	* Registers the dependencies of the ot of band module on the dependency manager.
	*/
	register(dependencyManager) {
		dependencyManager.registerSingleton(DidCommOutOfBandService);
		dependencyManager.registerSingleton(DidCommOutOfBandRepository);
	}
	async initialize(agentContext) {
		const featureRegistry = agentContext.resolve(DidCommFeatureRegistry);
		const messageHandlerRegistry = agentContext.resolve(DidCommMessageHandlerRegistry);
		const outOfBandService = agentContext.resolve(DidCommOutOfBandService);
		messageHandlerRegistry.registerMessageHandler(new DidCommHandshakeReuseHandler(outOfBandService));
		messageHandlerRegistry.registerMessageHandler(new DidCommHandshakeReuseAcceptedHandler(outOfBandService));
		featureRegistry.register(new DidCommProtocol({
			id: "https://didcomm.org/out-of-band/1.1",
			roles: ["sender", "receiver"]
		}));
	}
};

//#endregion
export { DidCommOutOfBandModule };
