// @credo-ts/didcomm/build/modules/routing/messages/DidCommKeylistUpdateMessage.mjs
import { __decorate } from "../../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorate.mjs";
import { IsValidMessageType, parseMessageType } from "../../../util/messageType.mjs";
import { __decorateMetadata } from "../../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorateMetadata.mjs";
import { DidCommMessage } from "../../../DidCommMessage.mjs";
import { IsArray, IsEnum, IsInstance, IsString, ValidateNested } from "class-validator";
import { Expose, Type } from "class-transformer";

//#region src/modules/routing/messages/DidCommKeylistUpdateMessage.ts
var _ref;
let DidCommKeylistUpdateAction = /* @__PURE__ */ function(DidCommKeylistUpdateAction) {
	DidCommKeylistUpdateAction["add"] = "add";
	DidCommKeylistUpdateAction["remove"] = "remove";
	return DidCommKeylistUpdateAction;
}({});
var DidCommKeylistUpdate = class {
	constructor(options) {
		if (options) {
			this.recipientKey = options.recipientKey;
			this.action = options.action;
		}
	}
};
__decorate([
	IsString(),
	Expose({ name: "recipient_key" }),
	__decorateMetadata("design:type", String)
], DidCommKeylistUpdate.prototype, "recipientKey", void 0);
__decorate([IsEnum(DidCommKeylistUpdateAction), __decorateMetadata("design:type", typeof (_ref = typeof DidCommKeylistUpdateAction !== "undefined" && DidCommKeylistUpdateAction) === "function" ? _ref : Object)], DidCommKeylistUpdate.prototype, "action", void 0);
/**
* Used to notify the mediator of keys in use by the recipient.
*
* @see https://github.com/hyperledger/aries-rfcs/blob/master/features/0211-route-coordination/README.md#keylist-update
*/
var DidCommKeylistUpdateMessage = class DidCommKeylistUpdateMessage extends DidCommMessage {
	constructor(options) {
		super();
		this.type = DidCommKeylistUpdateMessage.type.messageTypeUri;
		if (options) {
			this.id = options.id || this.generateId();
			this.updates = options.updates;
		}
	}
};
DidCommKeylistUpdateMessage.type = parseMessageType("https://didcomm.org/coordinate-mediation/1.0/keylist-update");
__decorate([IsValidMessageType(DidCommKeylistUpdateMessage.type), __decorateMetadata("design:type", Object)], DidCommKeylistUpdateMessage.prototype, "type", void 0);
__decorate([
	Type(() => DidCommKeylistUpdate),
	IsArray(),
	ValidateNested(),
	IsInstance(DidCommKeylistUpdate, { each: true }),
	__decorateMetadata("design:type", Array)
], DidCommKeylistUpdateMessage.prototype, "updates", void 0);

//#endregion
export { DidCommKeylistUpdate, DidCommKeylistUpdateAction, DidCommKeylistUpdateMessage };
