// @credo-ts/didcomm/build/modules/routing/messages/DidCommForwardMessage.mjs
import { __decorate } from "../../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorate.mjs";
import { IsValidMessageType, parseMessageType } from "../../../util/messageType.mjs";
import { __decorateMetadata } from "../../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorateMetadata.mjs";
import { DidCommMessage } from "../../../DidCommMessage.mjs";
import { IsObject, IsString } from "class-validator";
import { Expose } from "class-transformer";

//#region src/modules/routing/messages/DidCommForwardMessage.ts
/**
* @see https://github.com/hyperledger/aries-rfcs/blob/master/concepts/0094-cross-domain-messaging/README.md#corerouting10forward
*/
var DidCommForwardMessage = class DidCommForwardMessage extends DidCommMessage {
	/**
	* Create new ForwardMessage instance.
	*
	* @param options
	*/
	constructor(options) {
		super();
		this.allowDidSovPrefix = true;
		this.type = DidCommForwardMessage.type.messageTypeUri;
		if (options) {
			this.id = options.id || this.generateId();
			this.to = options.to;
			this.message = options.message;
		}
	}
};
DidCommForwardMessage.type = parseMessageType("https://didcomm.org/routing/1.0/forward");
__decorate([IsValidMessageType(DidCommForwardMessage.type), __decorateMetadata("design:type", Object)], DidCommForwardMessage.prototype, "type", void 0);
__decorate([IsString(), __decorateMetadata("design:type", String)], DidCommForwardMessage.prototype, "to", void 0);
__decorate([
	Expose({ name: "msg" }),
	IsObject(),
	__decorateMetadata("design:type", Object)
], DidCommForwardMessage.prototype, "message", void 0);

//#endregion
export { DidCommForwardMessage };
