// @credo-ts/didcomm/build/DidCommModule.mjs
import { DidCommFeatureRegistry } from "./DidCommFeatureRegistry.mjs";
import { DidCommMessageHandlerRegistry } from "./DidCommMessageHandlerRegistry.mjs";
import { DidCommEventTypes } from "./DidCommEvents.mjs";
import { DidCommModuleConfig } from "./DidCommModuleConfig.mjs";
import { DidCommEnvelopeService } from "./DidCommEnvelopeService.mjs";
import { DidCommTransportService } from "./DidCommTransportService.mjs";
import { DidCommMessageSender } from "./DidCommMessageSender.mjs";
import { DidCommConnectionsModule } from "./modules/connections/DidCommConnectionsModule.mjs";
import { DidCommBasicMessagesModule } from "./modules/basic-messages/DidCommBasicMessagesModule.mjs";
import { DidCommOutOfBandModule } from "./modules/oob/DidCommOutOfBandModule.mjs";
import { DidCommDiscoverFeaturesModule } from "./modules/discover-features/DidCommDiscoverFeaturesModule.mjs";
import { DidCommMessagePickupModule } from "./modules/message-pickup/DidCommMessagePickupModule.mjs";
import { DidCommMediationRecipientModule } from "./modules/routing/DidCommMediationRecipientModule.mjs";
import { DidCommMediatorModule } from "./modules/routing/DidCommMediatorModule.mjs";
import { DidCommMessageRepository } from "./repository/DidCommMessageRepository.mjs";
import "./repository/index.mjs";
import { DidCommCredentialsModule } from "./modules/credentials/DidCommCredentialsModule.mjs";
import { DidCommProofsModule } from "./modules/proofs/DidCommProofsModule.mjs";
import "./modules/index.mjs";
import { DidCommApi } from "./DidCommApi.mjs";
import { DidCommDispatcher } from "./DidCommDispatcher.mjs";
import { DidCommMessageReceiver } from "./DidCommMessageReceiver.mjs";
import { updateV0_1ToV0_2 } from "./updates/0.1-0.2/index.mjs";
import { updateV0_2ToV0_3 } from "./updates/0.2-0.3/index.mjs";
import { updateV0_4ToV0_5 } from "./updates/0.4-0.5/index.mjs";
import { EventEmitter, InjectionSymbols } from "@credo-ts/core";
import { mergeMap, takeUntil } from "rxjs";

//#region src/DidCommModule.ts
function getDidcommModules(options) {
	return Object.fromEntries(Object.entries({
		connections: new DidCommConnectionsModule(options.connections),
		oob: new DidCommOutOfBandModule(),
		discovery: new DidCommDiscoverFeaturesModule(options.discovery),
		credentials: options.credentials !== false ? new DidCommCredentialsModule(options.credentials === true ? {} : options.credentials) : void 0,
		proofs: options.proofs !== false ? new DidCommProofsModule(options.proofs === true ? {} : options.proofs) : void 0,
		mediator: options.mediator !== false ? new DidCommMediatorModule(options.mediator === true ? {} : options.mediator) : void 0,
		mediationRecipient: options.mediationRecipient !== false ? new DidCommMediationRecipientModule(options.mediationRecipient === true ? {} : options.mediationRecipient) : void 0,
		messagePickup: options.messagePickup !== false ? new DidCommMessagePickupModule(options.messagePickup === true ? {} : options.messagePickup) : void 0,
		basicMessages: options.basicMessages !== false ? new DidCommBasicMessagesModule() : void 0
	}).filter(([, moduleValue]) => moduleValue !== void 0));
}
var DidCommModule = class {
	constructor(config) {
		this.api = DidCommApi;
		this.config = new DidCommModuleConfig(config);
		this.modules = getDidcommModules(config ?? {});
	}
	/**
	* Registers the dependencies of the question answer module on the dependency manager.
	*/
	register(dependencyManager) {
		dependencyManager.registerInstance(DidCommModuleConfig, this.config);
		dependencyManager.registerSingleton(DidCommMessageHandlerRegistry);
		dependencyManager.registerSingleton(DidCommFeatureRegistry);
		dependencyManager.registerSingleton(DidCommMessageSender);
		dependencyManager.registerSingleton(DidCommMessageReceiver);
		dependencyManager.registerSingleton(DidCommTransportService);
		dependencyManager.registerSingleton(DidCommDispatcher);
		dependencyManager.registerSingleton(DidCommEnvelopeService);
		dependencyManager.registerSingleton(DidCommMessageRepository);
		for (const [_moduleKey, module] of Object.entries(this.modules)) {
			module.register(dependencyManager);
			if (module.api) dependencyManager.registerContextScoped(module.api);
		}
	}
	async initialize(agentContext) {
		const stop$ = agentContext.dependencyManager.resolve(InjectionSymbols.Stop$);
		const eventEmitter = agentContext.dependencyManager.resolve(EventEmitter);
		const messageReceiver = agentContext.dependencyManager.resolve(DidCommMessageReceiver);
		eventEmitter.observable(DidCommEventTypes.DidCommMessageReceived).pipe(takeUntil(stop$), mergeMap((e) => messageReceiver.receiveMessage(e.payload.message, {
			connection: e.payload.connection,
			contextCorrelationId: e.payload.contextCorrelationId,
			session: e.payload.session,
			receivedAt: e.payload.receivedAt
		}).catch((error) => {
			agentContext.config.logger.error("Failed to process message", { error });
		}), this.config.processDidCommMessagesConcurrently ? void 0 : 1)).subscribe();
		for (const transport of this.config.inboundTransports) await transport.start(agentContext);
		for (const transport of this.config.outboundTransports) await transport.start(agentContext);
		for (const module of Object.values(this.modules)) await module.initialize?.(agentContext);
	}
	async onInitializeContext(agentContext) {
		for (const module of Object.values(this.modules)) await module.onInitializeContext?.(agentContext);
	}
	async onCloseContext(agentContext) {
		for (const module of Object.values(this.modules)) await module.onCloseContext?.(agentContext);
	}
	async onDeleteContext(agentContext) {
		for (const module of Object.values(this.modules)) await module.onDeleteContext?.(agentContext);
	}
	async onProvisionContext(agentContext) {
		for (const module of Object.values(this.modules)) await module.onProvisionContext?.(agentContext);
	}
	async shutdown(agentContext) {
		const transportPromises = [...this.config.inboundTransports, ...this.config.outboundTransports].map((transport) => transport.stop());
		await Promise.all(transportPromises);
		for (const module of Object.values(this.modules)) await module.shutdown?.(agentContext);
	}
	get updates() {
		return [
			{
				fromVersion: "0.1",
				toVersion: "0.2",
				doUpdate: updateV0_1ToV0_2
			},
			{
				fromVersion: "0.2",
				toVersion: "0.3",
				doUpdate: updateV0_2ToV0_3
			},
			{
				fromVersion: "0.4",
				toVersion: "0.5",
				doUpdate: updateV0_4ToV0_5
			}
		];
	}
};

//#endregion
export { DidCommModule };
