// @credo-ts/didcomm/build/DidCommModuleConfig.mjs
import { DID_COMM_TRANSPORT_QUEUE } from "./constants.mjs";
import { InMemoryQueueTransportRepository } from "./transport/queue/InMemoryQueueTransportRepository.mjs";
import "./transport/index.mjs";
import { DidCommMimeType } from "./types.mjs";

//#region src/DidCommModuleConfig.ts
var DidCommModuleConfig = class {
	constructor(options) {
		this.options = options ?? {};
		this._endpoints = options?.endpoints;
		this._inboundTransports = options?.transports?.inbound ?? [];
		this._outboundTransports = options?.transports?.outbound ?? [];
		this._queueTransportRepository = options?.queueTransportRepository ?? new InMemoryQueueTransportRepository();
		this.enabledModules = {
			connections: true,
			oob: true,
			discovery: true,
			proofs: this.options.proofs !== false,
			credentials: this.options.credentials !== false,
			messagePickup: this.options.messagePickup !== false,
			mediator: this.options.mediator !== false,
			mediationRecipient: this.options.mediationRecipient !== false,
			basicMessages: this.options.basicMessages !== false
		};
	}
	get endpoints() {
		if (!this._endpoints || this._endpoints.length === 0) return [DID_COMM_TRANSPORT_QUEUE];
		return this._endpoints;
	}
	set endpoints(endpoints) {
		this._endpoints = endpoints;
	}
	get useDidSovPrefixWhereAllowed() {
		return this.options.useDidSovPrefixWhereAllowed ?? false;
	}
	get inboundTransports() {
		return this._inboundTransports;
	}
	set inboundTransports(inboundTransports) {
		this._inboundTransports = inboundTransports;
	}
	get outboundTransports() {
		return this._outboundTransports;
	}
	set outboundTransports(outboundTransports) {
		this._outboundTransports = outboundTransports;
	}
	get processDidCommMessagesConcurrently() {
		return this.options.processDidCommMessagesConcurrently ?? false;
	}
	get didCommMimeType() {
		return this.options.didCommMimeType ?? DidCommMimeType.V1;
	}
	/**
	* Encode keys in did:key format instead of 'naked' keys, as stated in Aries RFC 0360.
	*
	* This setting will not be taken into account if the other party has previously used naked keys
	* in a given protocol (i.e. it does not support Aries RFC 0360).
	*/
	get useDidKeyInProtocols() {
		return this.options.useDidKeyInProtocols ?? true;
	}
	/**
	* Allows to specify a custom queue transport queue. It defaults to an in-memory queue
	*
	*/
	get queueTransportRepository() {
		return this._queueTransportRepository;
	}
};

//#endregion
export { DidCommModuleConfig };
