// @credo-ts/didcomm/build/DidCommApi.mjs
import { __decorate } from "./_virtual/_@oxc-project_runtime@0.110.0/helpers/decorate.mjs";
import { DidCommFeatureRegistry } from "./DidCommFeatureRegistry.mjs";
import { DidCommMessageHandlerRegistry } from "./DidCommMessageHandlerRegistry.mjs";
import { __decorateMetadata } from "./_virtual/_@oxc-project_runtime@0.110.0/helpers/decorateMetadata.mjs";
import { DidCommModuleConfig } from "./DidCommModuleConfig.mjs";
import { DidCommConnectionsApi } from "./modules/connections/DidCommConnectionsApi.mjs";
import { DidCommBasicMessagesApi } from "./modules/basic-messages/DidCommBasicMessagesApi.mjs";
import { DidCommOutOfBandApi } from "./modules/oob/DidCommOutOfBandApi.mjs";
import { DidCommDiscoverFeaturesApi } from "./modules/discover-features/DidCommDiscoverFeaturesApi.mjs";
import { DidCommMessagePickupApi } from "./modules/message-pickup/DidCommMessagePickupApi.mjs";
import { DidCommMediationRecipientApi } from "./modules/routing/DidCommMediationRecipientApi.mjs";
import { DidCommMediatorApi } from "./modules/routing/DidCommMediatorApi.mjs";
import { DidCommCredentialsApi } from "./modules/credentials/DidCommCredentialsApi.mjs";
import { DidCommProofsApi } from "./modules/proofs/DidCommProofsApi.mjs";
import "./modules/index.mjs";
import { AgentContext, injectable } from "@credo-ts/core";

//#region src/DidCommApi.ts
var _ref, _ref2, _ref3, _ref4;
let DidCommApi = class DidCommApi {
	constructor(agentContext, messageHandlerRegistry, featureRegistry, config) {
		this.agentContext = agentContext;
		this.messageHandlerRegistry = messageHandlerRegistry;
		this.featureRegistry = featureRegistry;
		this.config = config;
		this.connections = this.agentContext.resolve(DidCommConnectionsApi);
		this.oob = this.agentContext.resolve(DidCommOutOfBandApi);
		this.discovery = this.agentContext.resolve(DidCommDiscoverFeaturesApi);
		this.proofs = this.apiOrUndefined(this.config.enabledModules.proofs, DidCommProofsApi);
		this.credentials = this.apiOrUndefined(this.config.enabledModules.credentials, DidCommCredentialsApi);
		this.messagePickup = this.apiOrUndefined(this.config.enabledModules.messagePickup, DidCommMessagePickupApi);
		this.basicMessages = this.apiOrUndefined(this.config.enabledModules.basicMessages, DidCommBasicMessagesApi);
		this.mediator = this.apiOrUndefined(this.config.enabledModules.mediator, DidCommMediatorApi);
		this.mediationRecipient = this.apiOrUndefined(this.config.enabledModules.mediationRecipient, DidCommMediationRecipientApi);
	}
	registerInboundTransport(inboundTransport) {
		this.config.inboundTransports.push(inboundTransport);
	}
	async unregisterInboundTransport(inboundTransport) {
		this.config.inboundTransports = this.config.inboundTransports.filter((transport) => transport !== inboundTransport);
		await inboundTransport.stop();
	}
	get inboundTransports() {
		return this.config.inboundTransports;
	}
	registerOutboundTransport(outboundTransport) {
		this.config.outboundTransports.push(outboundTransport);
	}
	async unregisterOutboundTransport(outboundTransport) {
		this.config.outboundTransports = this.config.outboundTransports.filter((transport) => transport !== outboundTransport);
	}
	get outboundTransports() {
		return this.config.outboundTransports;
	}
	/**
	* Agent's feature registry
	*/
	registerMessageHandlers(messageHandlers) {
		for (const messageHandler of messageHandlers) this.messageHandlerRegistry.registerMessageHandler(messageHandler);
	}
	registerMessageHandlerMiddleware(messageHandlerMiddleware) {
		this.messageHandlerRegistry.registerMessageHandlerMiddleware(messageHandlerMiddleware);
	}
	get fallbackMessageHandler() {
		return this.messageHandlerRegistry.fallbackMessageHandler;
	}
	get messageHandlerMiddlewares() {
		return this.messageHandlerRegistry.messageHandlerMiddlewares;
	}
	/**
	* Sets the fallback message handler, the message handler that will be called if no handler
	* is registered for an incoming message type.
	*/
	setFallbackMessageHandler(fallbackMessageHandler) {
		this.messageHandlerRegistry.setFallbackMessageHandler(fallbackMessageHandler);
	}
	get features() {
		return this.featureRegistry;
	}
	apiOrUndefined(isEnabled, apiClass) {
		if (isEnabled) return this.agentContext.resolve(apiClass);
	}
};
DidCommApi = __decorate([injectable(), __decorateMetadata("design:paramtypes", [
	typeof (_ref = typeof AgentContext !== "undefined" && AgentContext) === "function" ? _ref : Object,
	typeof (_ref2 = typeof DidCommMessageHandlerRegistry !== "undefined" && DidCommMessageHandlerRegistry) === "function" ? _ref2 : Object,
	typeof (_ref3 = typeof DidCommFeatureRegistry !== "undefined" && DidCommFeatureRegistry) === "function" ? _ref3 : Object,
	typeof (_ref4 = typeof DidCommModuleConfig !== "undefined" && DidCommModuleConfig) === "function" ? _ref4 : Object
])], DidCommApi);

//#endregion
export { DidCommApi };
