// @credo-ts/core/build/modules/dids/methods/web/WebDidResolver.mjs


import { JsonTransformer } from "../../../../utils/JsonTransformer.mjs";
import { DidDocument } from "../../domain/DidDocument.mjs";
import "../../domain/index.mjs";
import { Resolver } from "did-resolver";
import * as didWeb from "web-did-resolver";

//#region src/modules/dids/methods/web/WebDidResolver.ts
var WebDidResolver = class {
	constructor() {
		this.allowsCaching = true;
		this.allowsLocalDidRecord = true;
		this._resolverInstance = new Resolver();
		this.resolver = didWeb.getResolver();
		this.supportedMethods = Object.keys(this.resolver);
	}
	async resolve(_agentContext, did, parsed, didResolutionOptions) {
		const result = await this.resolver[parsed.method](did, parsed, this._resolverInstance, didResolutionOptions);
		let didDocument = null;
		if (!result.didDocument?.verificationMethod && result.didDocument?.publicKey) result.didDocument.verificationMethod = result.didDocument.publicKey;
		if (result.didDocument) didDocument = JsonTransformer.fromJSON(result.didDocument, DidDocument);
		return {
			...result,
			didDocument
		};
	}
};

//#endregion
export { WebDidResolver };
