// @credo-ts/didcomm/build/updates/0.2-0.3/proof.mjs
import { DidCommMessageRecord } from "../../repository/DidCommMessageRecord.mjs";
import { DidCommMessageRepository } from "../../repository/DidCommMessageRepository.mjs";
import { DidCommMessageRole } from "../../repository/DidCommMessageRole.mjs";
import "../../repository/index.mjs";
import { DidCommProofState } from "../../modules/proofs/models/DidCommProofState.mjs";
import { DidCommProofExchangeRepository } from "../../modules/proofs/repository/DidCommProofExchangeRepository.mjs";
import "../../modules/proofs/index.mjs";

//#region src/updates/0.2-0.3/proof.ts
/**
* Migrates the {@link DidCommProofExchangeRecord} to 0.3 compatible format. It fetches all records from storage
* and applies the needed updates to the records. After a record has been transformed, it is updated
* in storage and the next record will be transformed.
*
* The following transformations are applied:
*  - {@link migrateInternalProofExchangeRecordProperties}
*  - {@link moveDidCommMessages}
*/
async function migrateProofExchangeRecordToV0_3(agent) {
	agent.config.logger.info("Migrating proof records to storage version 0.3");
	const proofRepository = agent.dependencyManager.resolve(DidCommProofExchangeRepository);
	agent.config.logger.debug("Fetching all proof records from storage");
	const allProofs = await proofRepository.getAll(agent.context);
	agent.config.logger.debug(`Found a total of ${allProofs.length} proof records to update.`);
	for (const proofRecord of allProofs) {
		agent.config.logger.debug(`Migrating proof record with id ${proofRecord.id} to storage version 0.3`);
		await migrateInternalProofExchangeRecordProperties(agent, proofRecord);
		await moveDidCommMessages(agent, proofRecord);
		await proofRepository.update(agent.context, proofRecord);
		agent.config.logger.debug(`Successfully migrated proof record with id ${proofRecord.id} to storage version 0.3`);
	}
}
let V02_03MigrationProofRole = /* @__PURE__ */ function(V02_03MigrationProofRole) {
	V02_03MigrationProofRole[V02_03MigrationProofRole["Verifier"] = 0] = "Verifier";
	V02_03MigrationProofRole[V02_03MigrationProofRole["Prover"] = 1] = "Prover";
	return V02_03MigrationProofRole;
}({});
const proverProofStates = [
	DidCommProofState.Declined,
	DidCommProofState.ProposalSent,
	DidCommProofState.RequestReceived,
	DidCommProofState.PresentationSent,
	DidCommProofState.Done
];
const didCommMessageRoleMapping = {
	[V02_03MigrationProofRole.Verifier]: {
		proposalMessage: DidCommMessageRole.Receiver,
		requestMessage: DidCommMessageRole.Sender,
		presentationMessage: DidCommMessageRole.Receiver
	},
	[V02_03MigrationProofRole.Prover]: {
		proposalMessage: DidCommMessageRole.Sender,
		requestMessage: DidCommMessageRole.Receiver,
		presentationMessage: DidCommMessageRole.Sender
	}
};
const proofRecordMessageKeys = [
	"proposalMessage",
	"requestMessage",
	"presentationMessage"
];
function getProofRole(proofRecord) {
	if (proofRecord.isVerified !== void 0) return V02_03MigrationProofRole.Verifier;
	if (proofRecord.state === DidCommProofState.Done) return V02_03MigrationProofRole.Prover;
	if (proverProofStates.includes(proofRecord.state)) return V02_03MigrationProofRole.Prover;
	return V02_03MigrationProofRole.Verifier;
}
/**
* With the addition of support for different protocol versions the proof record now stores the protocol version.
*
* The following 0.2.0 proof record structure (unrelated keys omitted):
*
* ```json
* {
* }
* ```
*
* Will be transformed into the following 0.3.0 structure (unrelated keys omitted):
*
* ```json
* {
*  "protocolVersion: "v1"
* }
* ```
*/
async function migrateInternalProofExchangeRecordProperties(agent, proofRecord) {
	agent.config.logger.debug(`Migrating internal proof record ${proofRecord.id} properties to storage version 0.3`);
	if (!proofRecord.protocolVersion) {
		agent.config.logger.debug("Setting protocolVersion to v1");
		proofRecord.protocolVersion = "v1";
	}
	agent.config.logger.debug(`Successfully migrated internal proof record ${proofRecord.id} properties to storage version 0.3`);
}
/**
* In 0.3.0 the v1 didcomm messages have been moved out of the proof record into separate record using the DidCommMessageRepository.
* This migration scripts extracts all message (proposalMessage, requestMessage, presentationMessage) and moves
* them into the DidCommMessageRepository.
*/
async function moveDidCommMessages(agent, proofRecord) {
	agent.config.logger.debug(`Moving didcomm messages from proof record with id ${proofRecord.id} to DidCommMessageRecord`);
	const didCommMessageRepository = agent.dependencyManager.resolve(DidCommMessageRepository);
	for (const messageKey of proofRecordMessageKeys) {
		agent.config.logger.debug(`Starting move of ${messageKey} from proof record with id ${proofRecord.id} to DIDCommMessageRecord`);
		const proofRecordJson = proofRecord;
		const message = proofRecordJson[messageKey];
		if (message) {
			const didCommMessageRole = didCommMessageRoleMapping[getProofRole(proofRecord)][messageKey];
			const didCommMessageRecord = new DidCommMessageRecord({
				role: didCommMessageRole,
				associatedRecordId: proofRecord.id,
				message
			});
			await didCommMessageRepository.save(agent.context, didCommMessageRecord);
			agent.config.logger.debug(`Successfully moved ${messageKey} from proof record with id ${proofRecord.id} to DIDCommMessageRecord`);
			delete proofRecordJson[messageKey];
		} else agent.config.logger.debug(`Proof record with id ${proofRecord.id} does not have a ${messageKey}. Not creating a DIDCommMessageRecord`);
	}
	agent.config.logger.debug(`Successfully moved didcomm messages from proof record with id ${proofRecord.id} to DIDCommMessageRecord`);
}

//#endregion
export { migrateProofExchangeRecordToV0_3 };
