// @animo-id/pex/dist/module/lib/evaluation/evaluationClient.js
import { IProofType } from '@sphereon/ssi-types';
import { Status } from '../ConstraintUtils';
import PexMessages from '../types/Messages';
import { filterToRestrictedDIDs, uniformDIDMethods } from '../utils';
import { DIDRestrictionEvaluationHandler, FormatRestrictionEvaluationHandler, InputDescriptorFilterEvaluationHandler, LimitDisclosureEvaluationHandler, MarkForSubmissionEvaluationHandler, PredicateRelatedFieldEvaluationHandler, SameSubjectEvaluationHandler, SubjectIsHolderEvaluationHandler, SubjectIsIssuerEvaluationHandler, UriEvaluationHandler, } from './handlers';
const DEFAULT_LIMIT_DISCLOSURE_TYPES = [
    IProofType.BbsBlsSignatureProof2020,
    'DataIntegrityProof.anoncredsvc-2023',
    'DataIntegrityProof.anoncreds-2023',
];
export class EvaluationClient {
    constructor() {
        this._results = [];
        this._wrappedVcs = [];
        this._presentationSubmission = {};
        this._dids = [];
        this._limitDisclosureSignatureSuites = DEFAULT_LIMIT_DISCLOSURE_TYPES;
        this._restrictToDIDMethods = [];
        this._generatePresentationSubmission = true;
    }
    failed_catched = {
        tag: 'root',
        status: Status.ERROR,
        message: PexMessages.UNKNOWN_EXCEPTION,
        stacktrace: '',
    };
    _results;
    _wrappedVcs;
    _presentationSubmission;
    // private _requirePresentationSubmission: boolean;
    _dids;
    _limitDisclosureSignatureSuites;
    _restrictToFormats;
    _restrictToDIDMethods;
    _generatePresentationSubmission;
    evaluate(pd, wvcs, opts) {
        this._restrictToDIDMethods = opts?.restrictToDIDMethods ? uniformDIDMethods(opts?.restrictToDIDMethods) : [];
        this._dids = opts?.holderDIDs ? filterToRestrictedDIDs(opts.holderDIDs, this._restrictToDIDMethods) : [];
        this._limitDisclosureSignatureSuites = opts?.limitDisclosureSignatureSuites;
        this._restrictToFormats = opts?.restrictToFormats;
        this._generatePresentationSubmission = opts?.generatePresentationSubmission !== undefined ? opts.generatePresentationSubmission : true;
        if (opts?.presentationSubmission) {
            this._presentationSubmission = opts.presentationSubmission;
        }
        let currentHandler = this.initEvaluationHandlers();
        currentHandler?.handle(pd, wvcs);
        while (currentHandler?.hasNext()) {
            currentHandler = currentHandler.getNext();
            try {
                currentHandler?.handle(pd, wvcs);
            }
            catch (e) {
                this.failed_catched.message += e.message;
                this.failed_catched.stacktrace = e;
                throw this.failed_catched;
            }
        }
        // filter the presentation submission
        this.presentationSubmission = {
            ...this.presentationSubmission,
            descriptor_map: this.presentationSubmission.descriptor_map.filter((d) => d),
        };
    }
    get results() {
        return this._results;
    }
    get dids() {
        return this._dids;
    }
    set dids(dids) {
        this._dids = dids;
    }
    assertPresentationSubmission() {
        if (typeof this._presentationSubmission === 'string') {
            console.log('Presentation submission present, but as string not object. External calls did not follow contract. Correcting');
            this._presentationSubmission = JSON.parse(this._presentationSubmission);
        }
        if (!this.generatePresentationSubmission && (!this.presentationSubmission || Object.keys(this.presentationSubmission).length === 0)) {
            throw Error('No presentation submission present, but required option was set');
        }
    }
    get generatePresentationSubmission() {
        return this._generatePresentationSubmission;
    }
    set generatePresentationSubmission(value) {
        this._generatePresentationSubmission = value;
    }
    get presentationSubmission() {
        return this._presentationSubmission;
    }
    set presentationSubmission(presentationSubmission) {
        this._presentationSubmission = presentationSubmission;
    }
    get wrappedVcs() {
        return this._wrappedVcs;
    }
    set wrappedVcs(wrappedVcs) {
        this._wrappedVcs = wrappedVcs;
    }
    get limitDisclosureSignatureSuites() {
        return this._limitDisclosureSignatureSuites || DEFAULT_LIMIT_DISCLOSURE_TYPES;
    }
    set limitDisclosureSignatureSuites(limitDisclosureSignatureSuites) {
        this._limitDisclosureSignatureSuites = limitDisclosureSignatureSuites;
    }
    get restrictToDIDMethods() {
        return this._restrictToDIDMethods;
    }
    set restrictToDIDMethods(value) {
        this._restrictToDIDMethods = uniformDIDMethods(value);
    }
    hasRestrictToDIDMethods() {
        return this.restrictToDIDMethods && this.restrictToDIDMethods.length > 0;
    }
    get restrictToFormats() {
        return this._restrictToFormats;
    }
    set restrictToFormats(value) {
        this._restrictToFormats = value;
    }
    initEvaluationHandlers() {
        const uriEvaluation = new UriEvaluationHandler(this);
        uriEvaluation
            .setNext(new DIDRestrictionEvaluationHandler(this))
            .setNext(new FormatRestrictionEvaluationHandler(this))
            .setNext(new InputDescriptorFilterEvaluationHandler(this))
            .setNext(new PredicateRelatedFieldEvaluationHandler(this))
            .setNext(new LimitDisclosureEvaluationHandler(this))
            .setNext(new SubjectIsIssuerEvaluationHandler(this))
            .setNext(new SubjectIsHolderEvaluationHandler(this))
            .setNext(new SameSubjectEvaluationHandler(this))
            .setNext(new MarkForSubmissionEvaluationHandler(this));
        return uriEvaluation;
    }
}
