// @credo-ts/core/build/logger/ConsoleLogger.mjs


import { LogLevel } from "./Logger.mjs";
import { BaseLogger } from "./BaseLogger.mjs";
import { replaceError } from "./replaceError.mjs";

//#region src/logger/ConsoleLogger.ts
var ConsoleLogger = class extends BaseLogger {
	constructor(..._args) {
		super(..._args);
		this.consoleLogMap = {
			[LogLevel.test]: "log",
			[LogLevel.trace]: "log",
			[LogLevel.debug]: "debug",
			[LogLevel.info]: "info",
			[LogLevel.warn]: "warn",
			[LogLevel.error]: "error",
			[LogLevel.fatal]: "error"
		};
	}
	log(level, message, data) {
		const consoleLevel = this.consoleLogMap[level];
		const prefix = LogLevel[level].toUpperCase();
		if (!this.isEnabled(level)) return;
		if (data) console[consoleLevel](`${prefix}: ${message}`, JSON.stringify(data, replaceError, 2));
		else console[consoleLevel](`${prefix}: ${message}`);
	}
	test(message, data) {
		this.log(LogLevel.test, message, data);
	}
	trace(message, data) {
		this.log(LogLevel.trace, message, data);
	}
	debug(message, data) {
		this.log(LogLevel.debug, message, data);
	}
	info(message, data) {
		this.log(LogLevel.info, message, data);
	}
	warn(message, data) {
		this.log(LogLevel.warn, message, data);
	}
	error(message, data) {
		this.log(LogLevel.error, message, data);
	}
	fatal(message, data) {
		this.log(LogLevel.fatal, message, data);
	}
};

//#endregion
export { ConsoleLogger };
