// @credo-ts/didcomm/build/decorators/timing/TimingDecorator.mjs
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorate.mjs";
import { __decorateMetadata } from "../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorateMetadata.mjs";
import { IsDate, IsNumber, IsOptional } from "class-validator";
import { Expose, Type } from "class-transformer";

//#region src/decorators/timing/TimingDecorator.ts
var _ref, _ref2, _ref3, _ref4, _ref5;
/**
* Represents `~timing` decorator
* @see https://github.com/hyperledger/aries-rfcs/blob/master/features/0032-message-timing/README.md
*/
var TimingDecorator = class {
	constructor(partial) {
		this.inTime = partial?.inTime;
		this.outTime = partial?.outTime;
		this.staleTime = partial?.staleTime;
		this.expiresTime = partial?.expiresTime;
		this.delayMilli = partial?.delayMilli;
		this.waitUntilTime = partial?.waitUntilTime;
	}
};
__decorate([
	Expose({ name: "in_time" }),
	Type(() => Date),
	IsDate(),
	IsOptional(),
	__decorateMetadata("design:type", typeof (_ref = typeof Date !== "undefined" && Date) === "function" ? _ref : Object)
], TimingDecorator.prototype, "inTime", void 0);
__decorate([
	Expose({ name: "out_time" }),
	Type(() => Date),
	IsDate(),
	IsOptional(),
	__decorateMetadata("design:type", typeof (_ref2 = typeof Date !== "undefined" && Date) === "function" ? _ref2 : Object)
], TimingDecorator.prototype, "outTime", void 0);
__decorate([
	Expose({ name: "stale_time" }),
	Type(() => Date),
	IsDate(),
	IsOptional(),
	__decorateMetadata("design:type", typeof (_ref3 = typeof Date !== "undefined" && Date) === "function" ? _ref3 : Object)
], TimingDecorator.prototype, "staleTime", void 0);
__decorate([
	Expose({ name: "expires_time" }),
	Type(() => Date),
	IsDate(),
	IsOptional(),
	__decorateMetadata("design:type", typeof (_ref4 = typeof Date !== "undefined" && Date) === "function" ? _ref4 : Object)
], TimingDecorator.prototype, "expiresTime", void 0);
__decorate([
	Expose({ name: "delay_milli" }),
	IsNumber(),
	IsOptional(),
	__decorateMetadata("design:type", Number)
], TimingDecorator.prototype, "delayMilli", void 0);
__decorate([
	Expose({ name: "wait_until_time" }),
	Type(() => Date),
	IsDate(),
	IsOptional(),
	__decorateMetadata("design:type", typeof (_ref5 = typeof Date !== "undefined" && Date) === "function" ? _ref5 : Object)
], TimingDecorator.prototype, "waitUntilTime", void 0);

//#endregion
export { TimingDecorator };
