// @credo-ts/didcomm/build/modules/basic-messages/DidCommBasicMessagesModule.mjs
import { DidCommFeatureRegistry } from "../../DidCommFeatureRegistry.mjs";
import { DidCommMessageHandlerRegistry } from "../../DidCommMessageHandlerRegistry.mjs";
import { DidCommBasicMessageRole } from "./DidCommBasicMessageRole.mjs";
import { DidCommProtocol } from "../../models/features/DidCommProtocol.mjs";
import "../../models/index.mjs";
import { DidCommBasicMessageRepository } from "./repository/DidCommBasicMessageRepository.mjs";
import "./repository/index.mjs";
import { DidCommBasicMessageService } from "./services/DidCommBasicMessageService.mjs";
import "./services/index.mjs";
import { DidCommBasicMessagesApi } from "./DidCommBasicMessagesApi.mjs";
import { DidCommBasicMessageHandler } from "./handlers/BasicMessageHandler.mjs";
import "./handlers/index.mjs";

//#region src/modules/basic-messages/DidCommBasicMessagesModule.ts
var DidCommBasicMessagesModule = class {
	constructor() {
		this.api = DidCommBasicMessagesApi;
	}
	/**
	* Registers the dependencies of the basic message module on the dependency manager.
	*/
	register(dependencyManager) {
		dependencyManager.registerSingleton(DidCommBasicMessageService);
		dependencyManager.registerSingleton(DidCommBasicMessageRepository);
	}
	async initialize(agentContext) {
		const featureRegistry = agentContext.dependencyManager.resolve(DidCommFeatureRegistry);
		const messageHandlerRegistry = agentContext.resolve(DidCommMessageHandlerRegistry);
		const basicMessageService = agentContext.resolve(DidCommBasicMessageService);
		messageHandlerRegistry.registerMessageHandler(new DidCommBasicMessageHandler(basicMessageService));
		featureRegistry.register(new DidCommProtocol({
			id: "https://didcomm.org/basicmessage/1.0",
			roles: [DidCommBasicMessageRole.Sender, DidCommBasicMessageRole.Receiver]
		}));
	}
};

//#endregion
export { DidCommBasicMessagesModule };
