// @credo-ts/didcomm/build/modules/proofs/protocol/v2/handlers/DidCommProposePresentationV2Handler.mjs
import { DidCommOutboundMessageContext } from "../../../../../models/DidCommOutboundMessageContext.mjs";
import "../../../../../models/index.mjs";
import { DidCommProposePresentationV2Message } from "../messages/DidCommProposePresentationV2Message.mjs";

//#region src/modules/proofs/protocol/v2/handlers/DidCommProposePresentationV2Handler.ts
var DidCommProposePresentationV2Handler = class {
	constructor(proofProtocol) {
		this.supportedMessages = [DidCommProposePresentationV2Message];
		this.proofProtocol = proofProtocol;
	}
	async handle(messageContext) {
		const proofRecord = await this.proofProtocol.processProposal(messageContext);
		if (await this.proofProtocol.shouldAutoRespondToProposal(messageContext.agentContext, {
			proofRecord,
			proposalMessage: messageContext.message
		})) return this.acceptProposal(proofRecord, messageContext);
	}
	async acceptProposal(proofRecord, messageContext) {
		messageContext.agentContext.config.logger.info("Automatically sending request with autoAccept");
		if (!messageContext.connection) {
			messageContext.agentContext.config.logger.error("No connection on the messageContext, aborting auto accept");
			return;
		}
		const { message } = await this.proofProtocol.acceptProposal(messageContext.agentContext, { proofRecord });
		return new DidCommOutboundMessageContext(message, {
			agentContext: messageContext.agentContext,
			connection: messageContext.connection,
			associatedRecord: proofRecord
		});
	}
};

//#endregion
export { DidCommProposePresentationV2Handler };
