// @credo-ts/didcomm/build/modules/proofs/protocol/v2/handlers/DidCommPresentationV2Handler.mjs
import { DidCommMessageRepository } from "../../../../../repository/DidCommMessageRepository.mjs";
import { DidCommMessageRole } from "../../../../../repository/DidCommMessageRole.mjs";
import "../../../../../repository/index.mjs";
import { getOutboundDidCommMessageContext } from "../../../../../getDidCommOutboundMessageContext.mjs";
import { DidCommPresentationV2Message } from "../messages/DidCommPresentationV2Message.mjs";
import { DidCommRequestPresentationV2Message } from "../messages/DidCommRequestPresentationV2Message.mjs";
import "../messages/index.mjs";

//#region src/modules/proofs/protocol/v2/handlers/DidCommPresentationV2Handler.ts
var DidCommPresentationV2Handler = class {
	constructor(proofProtocol) {
		this.supportedMessages = [DidCommPresentationV2Message];
		this.proofProtocol = proofProtocol;
	}
	async handle(messageContext) {
		const proofRecord = await this.proofProtocol.processPresentation(messageContext);
		if (await this.proofProtocol.shouldAutoRespondToPresentation(messageContext.agentContext, {
			proofRecord,
			presentationMessage: messageContext.message
		})) return await this.acceptPresentation(proofRecord, messageContext);
	}
	async acceptPresentation(proofRecord, messageContext) {
		messageContext.agentContext.config.logger.info("Automatically sending acknowledgement with autoAccept");
		const { message } = await this.proofProtocol.acceptPresentation(messageContext.agentContext, { proofRecord });
		const requestMessage = await messageContext.agentContext.dependencyManager.resolve(DidCommMessageRepository).getAgentMessage(messageContext.agentContext, {
			associatedRecordId: proofRecord.id,
			messageClass: DidCommRequestPresentationV2Message,
			role: DidCommMessageRole.Sender
		});
		return getOutboundDidCommMessageContext(messageContext.agentContext, {
			connectionRecord: messageContext.connection,
			message,
			associatedRecord: proofRecord,
			lastReceivedMessage: messageContext.message,
			lastSentMessage: requestMessage
		});
	}
};

//#endregion
export { DidCommPresentationV2Handler };
