// @credo-ts/didcomm/build/modules/proofs/protocol/v2/handlers/DidCommRequestPresentationV2Handler.mjs
import { getOutboundDidCommMessageContext } from "../../../../../getDidCommOutboundMessageContext.mjs";
import { DidCommRequestPresentationV2Message } from "../messages/DidCommRequestPresentationV2Message.mjs";

//#region src/modules/proofs/protocol/v2/handlers/DidCommRequestPresentationV2Handler.ts
var DidCommRequestPresentationV2Handler = class {
	constructor(proofProtocol) {
		this.supportedMessages = [DidCommRequestPresentationV2Message];
		this.proofProtocol = proofProtocol;
	}
	async handle(messageContext) {
		const proofRecord = await this.proofProtocol.processRequest(messageContext);
		const shouldAutoRespond = await this.proofProtocol.shouldAutoRespondToRequest(messageContext.agentContext, {
			proofRecord,
			requestMessage: messageContext.message
		});
		messageContext.agentContext.config.logger.debug(`Should auto respond to request: ${shouldAutoRespond}`);
		if (shouldAutoRespond) return await this.acceptRequest(proofRecord, messageContext);
	}
	async acceptRequest(proofRecord, messageContext) {
		messageContext.agentContext.config.logger.info("Automatically sending presentation with autoAccept");
		const { message } = await this.proofProtocol.acceptRequest(messageContext.agentContext, { proofRecord });
		return getOutboundDidCommMessageContext(messageContext.agentContext, {
			message,
			lastReceivedMessage: messageContext.message,
			associatedRecord: proofRecord,
			connectionRecord: messageContext.connection
		});
	}
};

//#endregion
export { DidCommRequestPresentationV2Handler };
