// @credo-ts/didcomm/build/modules/proofs/DidCommProofsModule.mjs
import { DidCommFeatureRegistry } from "../../DidCommFeatureRegistry.mjs";
import { DidCommMessageHandlerRegistry } from "../../DidCommMessageHandlerRegistry.mjs";
import { DidCommProofsModuleConfig } from "./DidCommProofsModuleConfig.mjs";
import { DidCommProofExchangeRepository } from "./repository/DidCommProofExchangeRepository.mjs";
import { DidCommProofsApi } from "./DidCommProofsApi.mjs";
import "./repository/index.mjs";
import { DidCommProofV2Protocol } from "./protocol/v2/DidCommProofV2Protocol.mjs";
import "./protocol/index.mjs";

//#region src/modules/proofs/DidCommProofsModule.ts
var DidCommProofsModule = class {
	constructor(config) {
		this.api = DidCommProofsApi;
		this.config = new DidCommProofsModuleConfig({
			...config,
			proofProtocols: config?.proofProtocols ?? [new DidCommProofV2Protocol({ proofFormats: [] })]
		});
	}
	/**
	* Registers the dependencies of the proofs module on the dependency manager.
	*/
	register(dependencyManager) {
		dependencyManager.registerInstance(DidCommProofsModuleConfig, this.config);
		dependencyManager.registerSingleton(DidCommProofExchangeRepository);
	}
	async initialize(agentContext) {
		const messageHandlerRegistry = agentContext.dependencyManager.resolve(DidCommMessageHandlerRegistry);
		const featureRegistry = agentContext.dependencyManager.resolve(DidCommFeatureRegistry);
		for (const proofProtocol of this.config.proofProtocols) {
			proofProtocol.register(messageHandlerRegistry, featureRegistry);
			if (proofProtocol.constructor.name === "DidCommCredentialV1Protocol") agentContext.config.logger.debug("The 'DidCommCredentialV1Protocol' is deprecated and will be removed in version 0.7 of Credo. You should upgrade to the 'DidCommCredentialV2Protocol' instead.");
		}
	}
};

//#endregion
export { DidCommProofsModule };
