// @credo-ts/core/build/storage/migration/updates.mjs


import { updateV0_3_1ToV0_4 } from "./updates/0.3.1-0.4/index.mjs";
import { updateV0_3ToV0_3_1 } from "./updates/0.3-0.3.1/index.mjs";
import { updateV0_4ToV0_5 } from "./updates/0.4-0.5/index.mjs";

//#region src/storage/migration/updates.ts
const INITIAL_STORAGE_VERSION = "0.1";
const DEFAULT_UPDATE_CONFIG = { v0_1ToV0_2: { mediationRoleUpdateStrategy: "recipientIfEndpoint" } };
const supportedUpdates = [
	{
		fromVersion: "0.1",
		toVersion: "0.2",
		doUpdate: () => {}
	},
	{
		fromVersion: "0.2",
		toVersion: "0.3",
		doUpdate: () => {}
	},
	{
		fromVersion: "0.3",
		toVersion: "0.3.1",
		doUpdate: updateV0_3ToV0_3_1
	},
	{
		fromVersion: "0.3.1",
		toVersion: "0.4",
		doUpdate: updateV0_3_1ToV0_4
	},
	{
		fromVersion: "0.4",
		toVersion: "0.5",
		doUpdate: updateV0_4ToV0_5
	}
];
const CURRENT_FRAMEWORK_STORAGE_VERSION = supportedUpdates[supportedUpdates.length - 1].toVersion;
const STORAGE_VERSION_RECORD_ID = "STORAGE_VERSION_RECORD_ID";

//#endregion
export { CURRENT_FRAMEWORK_STORAGE_VERSION, DEFAULT_UPDATE_CONFIG, INITIAL_STORAGE_VERSION, STORAGE_VERSION_RECORD_ID, supportedUpdates };
