// @animo-id/pex/dist/main/lib/types/PexCredentialMapper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMdocDecodedPayload = exports.PexCredentialMapper = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */
const mdoc_1 = require("@animo-id/mdoc");
const ssi_types_1 = require("@sphereon/ssi-types");
const u8a = __importStar(require("uint8arrays"));
// NOTE: this is partial reimplementation of CredentialMapper from ssi-types to overwrite the
// methods relying on Sphereon's mdoc implementation
class PexCredentialMapper {
    static isMsoMdocDecodedPresentation(original) {
        return original instanceof mdoc_1.MDoc;
    }
    static isMsoMdocDecodedCredential(original) {
        return original instanceof mdoc_1.IssuerSignedDocument;
    }
    /**
     * Decodes a Verifiable Presentation to a uniform format.
     *
     * When decoding SD-JWT credentials, a hasher implementation must be provided. The hasher implementation must be sync. When using
     * an async hasher implementation, use the decodeSdJwtVcAsync method instead and you can provide the decoded payload to methods
     * instead of the compact SD-JWT.
     *
     * @param presentation
     * @param hasher Hasher implementation to use for SD-JWT decoding.
     */
    static decodeVerifiablePresentation(presentation, hasher) {
        if (ssi_types_1.CredentialMapper.isMsoMdocOid4VPEncoded(presentation)) {
            return presentation;
        }
        else if (this.isMsoMdocDecodedPresentation(presentation)) {
            return presentation;
        }
        else {
            return ssi_types_1.CredentialMapper.decodeVerifiablePresentation(presentation, hasher);
        }
    }
    static isW3cCredential(credential) {
        return ssi_types_1.CredentialMapper.isW3cCredential(credential);
    }
    /**
     * Converts a presentation to a wrapped presentation.
     *
     * When decoding SD-JWT credentials, a hasher implementation must be provided. The hasher implementation must be sync. When using
     * an async hasher implementation, use the decodeSdJwtVcAsync method instead and you can provide the decoded payload to methods
     * instead of the compact SD-JWT.
     *
     * @param hasher Hasher implementation to use for SD-JWT decoding
     */
    static toWrappedVerifiablePresentation(originalPresentation, opts) {
        var _a;
        // MSO_MDOC
        if (this.isMsoMdocDecodedPresentation(originalPresentation) || ssi_types_1.CredentialMapper.isMsoMdocOid4VPEncoded(originalPresentation)) {
            let deviceResponse;
            let originalType;
            if (ssi_types_1.CredentialMapper.isMsoMdocOid4VPEncoded(originalPresentation)) {
                deviceResponse = (0, mdoc_1.parseDeviceResponse)(convertBase64urlToBinary(originalPresentation));
                originalType = ssi_types_1.OriginalType.MSO_MDOC_ENCODED;
            }
            else {
                deviceResponse = originalPresentation;
                originalType = ssi_types_1.OriginalType.MSO_MDOC_DECODED;
            }
            const mdocCredentials = (_a = deviceResponse.documents) === null || _a === void 0 ? void 0 : _a.map((doc) => PexCredentialMapper.toWrappedVerifiableCredential(doc, opts));
            if (!mdocCredentials || mdocCredentials.length === 0) {
                throw new Error('could not extract any mdoc credentials from mdoc device response');
            }
            return {
                type: originalType,
                format: 'mso_mdoc',
                original: originalPresentation,
                presentation: deviceResponse,
                decoded: deviceResponse,
                vcs: mdocCredentials,
            };
        }
        return ssi_types_1.CredentialMapper.toWrappedVerifiablePresentation(originalPresentation, opts);
    }
    /**
     * Converts a credential to a wrapped credential.
     *
     * When decoding SD-JWT credentials, a hasher implementation must be provided. The hasher implementation must be sync. When using
     * an async hasher implementation, use the decodeSdJwtVcAsync method instead and you can provide the decoded payload to methods
     * instead of the compact SD-JWT.
     *
     * @param hasher Hasher implementation to use for SD-JWT decoding
     */
    static toWrappedVerifiableCredential(verifiableCredential, opts) {
        // MSO_MDOC
        if (this.isMsoMdocDecodedCredential(verifiableCredential) || ssi_types_1.CredentialMapper.isMsoMdocOid4VPEncoded(verifiableCredential)) {
            let mdoc;
            if (ssi_types_1.CredentialMapper.isMsoMdocOid4VPEncoded(verifiableCredential)) {
                mdoc = (0, mdoc_1.parseIssuerSigned)(convertBase64urlToBinary(verifiableCredential));
            }
            else {
                mdoc = verifiableCredential;
            }
            return {
                type: ssi_types_1.CredentialMapper.isMsoMdocDecodedCredential(verifiableCredential)
                    ? ssi_types_1.OriginalType.MSO_MDOC_DECODED
                    : ssi_types_1.OriginalType.MSO_MDOC_ENCODED,
                format: 'mso_mdoc',
                original: verifiableCredential,
                credential: mdoc,
                decoded: getMdocDecodedPayload(mdoc),
            };
        }
        return ssi_types_1.CredentialMapper.toWrappedVerifiableCredential(verifiableCredential, opts);
    }
    static isW3cPresentation(presentation) {
        return ssi_types_1.CredentialMapper.isW3cPresentation(presentation);
    }
    static isSdJwtDecodedCredential(original) {
        return ssi_types_1.CredentialMapper.isSdJwtDecodedCredential(original);
    }
    static isJwtDecodedCredential(original) {
        return ssi_types_1.CredentialMapper.isJwtDecodedCredential(original);
    }
    static isSdJwtEncoded(original) {
        return ssi_types_1.CredentialMapper.isSdJwtEncoded(original);
    }
    static isJwtEncoded(original) {
        return ssi_types_1.CredentialMapper.isJwtEncoded(original);
    }
    static decodeVerifiableCredential(credential, hasher) {
        return ssi_types_1.CredentialMapper.decodeVerifiableCredential(credential, hasher);
    }
    static isCredential(original) {
        return ssi_types_1.CredentialMapper.isCredential(original);
    }
    static areOriginalVerifiableCredentialsEqual(firstOriginal, secondOriginal) {
        if (this.isMsoMdocDecodedCredential(firstOriginal) || this.isMsoMdocDecodedCredential(secondOriginal)) {
            return (0, mdoc_1.uint8ArrayToBase64Url)((0, mdoc_1.cborEncode)(firstOriginal)) === (0, mdoc_1.uint8ArrayToBase64Url)((0, mdoc_1.cborEncode)(secondOriginal));
        }
        return ssi_types_1.CredentialMapper.areOriginalVerifiableCredentialsEqual(firstOriginal, secondOriginal);
    }
}
exports.PexCredentialMapper = PexCredentialMapper;
PexCredentialMapper.isWrappedSdJwtVerifiableCredential = (vc) => ssi_types_1.CredentialMapper.isWrappedSdJwtVerifiableCredential(vc);
PexCredentialMapper.isWrappedSdJwtVerifiablePresentation = (vc) => ssi_types_1.CredentialMapper.isWrappedSdJwtVerifiablePresentation(vc);
PexCredentialMapper.isWrappedW3CVerifiableCredential = (vc) => ssi_types_1.CredentialMapper.isWrappedW3CVerifiableCredential(vc);
PexCredentialMapper.isWrappedW3CVerifiablePresentation = (vc) => ssi_types_1.CredentialMapper.isWrappedW3CVerifiablePresentation(vc);
PexCredentialMapper.isWrappedMdocCredential = (vc) => ssi_types_1.CredentialMapper.isWrappedMdocCredential(vc);
PexCredentialMapper.isWrappedMdocPresentation = (vc) => ssi_types_1.CredentialMapper.isWrappedMdocPresentation(vc);
function getMdocDecodedPayload(mdoc) {
    const namespaces = mdoc.issuerSigned.nameSpaces;
    const decodedPayload = {};
    for (const [namespace, items] of Array.from(namespaces.entries())) {
        decodedPayload[namespace] = items.reduce((acc, item) => (Object.assign(Object.assign({}, acc), { [item.elementIdentifier]: encodeMdocValue(item.elementValue) })), {});
    }
    return decodedPayload;
}
exports.getMdocDecodedPayload = getMdocDecodedPayload;
function encodeMdocValue(value) {
    if (typeof value === 'string' || typeof value === 'number' || typeof value === 'undefined' || value === null)
        return value;
    if (value instanceof Date || value instanceof mdoc_1.DateOnly)
        return value.toISOString();
    // TODO: we don't want undefined, but empty object might also not work?
    return {};
}
function convertBase64urlToBinary(data) {
    return u8a.fromString(data, 'base64url');
}
