// @animo-id/pex/dist/browser/lib/evaluation/handlers/inputDescriptorFilterEvaluationHandler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InputDescriptorFilterEvaluationHandler = void 0;
const jsonpath_1 = require("@astronautlabs/jsonpath");
const ajv_1 = __importDefault(require("ajv"));
const ajv_formats_1 = __importDefault(require("ajv-formats"));
const ConstraintUtils_1 = require("../../ConstraintUtils");
const Messages_1 = __importDefault(require("../../types/Messages"));
const utils_1 = require("../../utils");
const abstractEvaluationHandler_1 = require("./abstractEvaluationHandler");
const AJV_FIELD_FILTER = new ajv_1.default({
    verbose: false,
    code: { source: false, lines: true, esm: false },
    allowUnionTypes: true,
    allErrors: true,
    strict: false,
});
(0, ajv_formats_1.default)(AJV_FIELD_FILTER);
class InputDescriptorFilterEvaluationHandler extends abstractEvaluationHandler_1.AbstractEvaluationHandler {
    constructor(client) {
        super(client);
        InputDescriptorFilterEvaluationHandler.keepCacheSizeInCheck();
    }
    getName() {
        return 'FilterEvaluation';
    }
    handle(pd, wrappedVcs) {
        const fields = jsonpath_1.JSONPath.nodes(pd, '$..fields[*]');
        wrappedVcs.forEach((wvc, vcIndex) => {
            this.createNoFieldResults(pd, vcIndex, wvc);
            fields.forEach((field) => {
                const isPredicate = field.value.predicate !== undefined;
                let inputField = [];
                if (field.value.path) {
                    inputField = utils_1.JsonPathUtils.extractInputField(wvc.decoded, field.value.path);
                }
                let resultFound = false;
                for (const inputFieldKey of inputField) {
                    if (this.evaluateFilter(inputFieldKey, field.value)) {
                        resultFound = true;
                        const payload = { result: Object.assign({}, inputField[0]), valid: true, format: wvc.format };
                        this.getResults().push(Object.assign({}, this.createResultObject(jsonpath_1.JSONPath.stringify(field.path.slice(0, 3)), vcIndex, payload)));
                    }
                    else if (isPredicate &&
                        this.evaluateFilter(inputFieldKey, Object.assign(Object.assign({}, field.value), { filter: {
                                type: 'boolean',
                                const: true,
                            } }))) {
                        resultFound = true;
                        const payload = { result: Object.assign({}, inputField[0]), valid: true, format: wvc.format, predicate: true };
                        this.getResults().push(Object.assign({}, this.createResultObject(jsonpath_1.JSONPath.stringify(field.path.slice(0, 3)), vcIndex, payload, Messages_1.default.INPUT_CANDIDATE_PREDICATE_VALUE_IS_ELIGIBLE_FOR_PRESENTATION_SUBMISSION)));
                    }
                }
                const isOptional = 'optional' in field.value && field.value.optional;
                if (!resultFound) {
                    if (inputField.length) {
                        const payload = { result: Object.assign({}, inputField[0]), valid: false, format: wvc.format };
                        this.createResponse(field, vcIndex, payload, Messages_1.default.INPUT_CANDIDATE_FAILED_FILTER_EVALUATION);
                    }
                    else if (!isOptional) {
                        const payload = { valid: false, format: wvc.format };
                        this.createResponse(field, vcIndex, payload, Messages_1.default.INPUT_CANDIDATE_DOESNT_CONTAIN_PROPERTY);
                    }
                    else {
                        this.getResults().push(Object.assign({}, this.createResultObject(jsonpath_1.JSONPath.stringify(field.path.slice(0, 3)), vcIndex, { result: undefined, valid: true, format: wvc.format })));
                    }
                }
            });
        });
        this.updatePresentationSubmission(pd);
    }
    createNoFieldResults(pd, vcIndex, credential) {
        // PresentationDefinitionV2 is the common denominator
        const noFields = pd.input_descriptors
            .map((inDesc, index) => {
            return { index, inDesc };
        })
            .filter((el) => { var _a, _b, _c; return ((_a = el.inDesc.constraints) === null || _a === void 0 ? void 0 : _a.fields) === undefined || ((_c = (_b = el.inDesc.constraints) === null || _b === void 0 ? void 0 : _b.fields) === null || _c === void 0 ? void 0 : _c.length) === 0; });
        noFields.forEach((noField) => {
            const payload = { result: [], valid: true, format: credential.format };
            this.getResults().push(Object.assign({}, this.createResultObject(`$.input_descriptors[${noField.index}]`, vcIndex, payload)));
        });
    }
    createResponse(field, vcIndex, payload, message) {
        this.getResults().push(Object.assign(Object.assign({}, this.createResultObject(jsonpath_1.JSONPath.stringify(field.path.slice(0, 3)), vcIndex, payload)), { ['status']: ConstraintUtils_1.Status.ERROR, ['message']: message }));
    }
    createResultObject(path, vcIndex, payload, message) {
        return {
            input_descriptor_path: path,
            verifiable_credential_path: `$[${vcIndex}]`,
            evaluator: this.getName(),
            status: ConstraintUtils_1.Status.INFO,
            message: message !== null && message !== void 0 ? message : Messages_1.default.INPUT_CANDIDATE_IS_ELIGIBLE_FOR_PRESENTATION_SUBMISSION,
            payload,
        };
    }
    evaluateFilter(result, field) {
        var _a, _b;
        if (((_a = field.filter) === null || _a === void 0 ? void 0 : _a.format) && field.filter.format === 'date') {
            this.transformDateFormat(result);
        }
        let evalResult = true;
        if (field.filter) {
            const successCacheKey = JSON.stringify({ filter: field.filter, value: result.value });
            const now = Date.now();
            evalResult = (_b = InputDescriptorFilterEvaluationHandler.FILTER_CACHE.get(successCacheKey)) === null || _b === void 0 ? void 0 : _b.value;
            if (evalResult === undefined) {
                InputDescriptorFilterEvaluationHandler.keepCacheSizeInCheck();
                evalResult = AJV_FIELD_FILTER.validate(field.filter, result.value);
                InputDescriptorFilterEvaluationHandler.FILTER_CACHE.set(successCacheKey, {
                    value: evalResult,
                    ts: now + InputDescriptorFilterEvaluationHandler.DEFAULT_FILTER_CACHE_TTL,
                });
            }
        }
        return evalResult;
    }
    transformDateFormat(result) {
        const date = new Date(result.value);
        let month = date.getUTCMonth() + 1 + '';
        if (month.length === 1) {
            month = '0' + month;
        }
        let day = date.getUTCDate() + '';
        if (day.length === 1) {
            day = '0' + day;
        }
        result.value = date.getUTCFullYear() + '-' + month + '-' + day;
        result.value = date.toISOString().substring(0, date.toISOString().indexOf('T'));
    }
    static keepCacheSizeInCheck(opts) {
        var _a, _b, _c;
        const ttl = (_a = opts === null || opts === void 0 ? void 0 : opts.ttl) !== null && _a !== void 0 ? _a : InputDescriptorFilterEvaluationHandler.DEFAULT_FILTER_CACHE_TTL;
        const maxCacheSize = (_b = opts === null || opts === void 0 ? void 0 : opts.maxCacheSize) !== null && _b !== void 0 ? _b : InputDescriptorFilterEvaluationHandler.DEFAULT_MAX_FILTER_CACHE_SIZE;
        const resetCacheSize = (_c = opts === null || opts === void 0 ? void 0 : opts.resetCacheSize) !== null && _c !== void 0 ? _c : InputDescriptorFilterEvaluationHandler.DEFAULT_RESET_CACHE_SIZE;
        const now = Date.now();
        for (const [key, result] of InputDescriptorFilterEvaluationHandler.FILTER_CACHE) {
            if (result.ts + ttl < now) {
                InputDescriptorFilterEvaluationHandler.FILTER_CACHE.delete(key);
            }
        }
        const size = InputDescriptorFilterEvaluationHandler.FILTER_CACHE.size;
        if (size > maxCacheSize) {
            const keys = InputDescriptorFilterEvaluationHandler.FILTER_CACHE.keys();
            // Since we cannot use a WeakMap, as the key is constructed on the fly and thus has no references, we need to clear the cache to avoid memory leaks
            for (let nr = 0; nr < size - resetCacheSize; nr++) {
                // would be wise to have sorted on oldest ts
                InputDescriptorFilterEvaluationHandler.FILTER_CACHE.delete(keys.next().value);
            }
        }
    }
}
exports.InputDescriptorFilterEvaluationHandler = InputDescriptorFilterEvaluationHandler;
InputDescriptorFilterEvaluationHandler.FILTER_CACHE = new Map();
InputDescriptorFilterEvaluationHandler.DEFAULT_MAX_FILTER_CACHE_SIZE = 100;
InputDescriptorFilterEvaluationHandler.DEFAULT_FILTER_CACHE_TTL = 1000 * 30; // 30 seconds
InputDescriptorFilterEvaluationHandler.DEFAULT_RESET_CACHE_SIZE = InputDescriptorFilterEvaluationHandler.DEFAULT_MAX_FILTER_CACHE_SIZE / 2;
