// @animo-id/pex/dist/browser/lib/evaluation/handlers/predicateRelatedFieldEvaluationHandler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PredicateRelatedFieldEvaluationHandler = void 0;
const pex_models_1 = require("@sphereon/pex-models");
const ConstraintUtils_1 = require("../../ConstraintUtils");
const Messages_1 = __importDefault(require("../../types/Messages"));
const abstractEvaluationHandler_1 = require("./abstractEvaluationHandler");
class PredicateRelatedFieldEvaluationHandler extends abstractEvaluationHandler_1.AbstractEvaluationHandler {
    constructor(client) {
        super(client);
    }
    getName() {
        return 'PredicateRelatedFieldEvaluation';
    }
    handle(pd) {
        // PresentationDefinitionV2 is the common denominator
        pd.input_descriptors.forEach((inDesc, index) => {
            if (inDesc.constraints) {
                this.examinePredicateRelatedField(index, inDesc.constraints);
            }
        });
        this.updatePresentationSubmission(pd);
    }
    examinePredicateRelatedField(input_descriptor_idx, constraints) {
        if (constraints === null || constraints === void 0 ? void 0 : constraints.fields) {
            for (let i = 0; i < constraints.fields.length; i++) {
                for (let j = 0; j < this.getResults().length; j++) {
                    this.examinePredicateForFilterEvaluationResult(this.getResults(), j, input_descriptor_idx, constraints, i);
                }
            }
        }
    }
    examinePredicateForFilterEvaluationResult(results, resultIdx, input_descriptor_idx, constraints, fieldIdx) {
        var _a, _b, _c;
        const resultInputDescriptorIdx = this.retrieveResultInputDescriptorIdx(results[resultIdx].input_descriptor_path);
        if (results[resultIdx].payload &&
            results[resultIdx].payload.result &&
            results[resultIdx].payload.result.path &&
            results[resultIdx].evaluator === 'FilterEvaluation' &&
            input_descriptor_idx === resultInputDescriptorIdx &&
            constraints &&
            constraints.fields &&
            constraints.fields[fieldIdx] &&
            constraints.fields[fieldIdx].predicate &&
            constraints.fields[fieldIdx].path &&
            ((_a = constraints.fields[fieldIdx].path) === null || _a === void 0 ? void 0 : _a.includes(this.concatenatePath(results[resultIdx].payload.result.path)))) {
            const field = constraints.fields[fieldIdx];
            const evaluationResult = Object.assign({}, results[resultIdx].payload.result);
            // We only support number with minimum/maximum for predicate type
            if ((((_b = field.filter) === null || _b === void 0 ? void 0 : _b.type) !== 'number' && ((_c = field.filter) === null || _c === void 0 ? void 0 : _c.type) !== 'integer') ||
                (!field.filter.minimum && !field.filter.exclusiveMinimum && !field.filter.maximum && !field.filter.exclusiveMaximum)) {
                results.push(this.createErrorResultObject(input_descriptor_idx, resultIdx, evaluationResult, results, "Only 'number' and 'integer' predicate with 'minimum', 'exclusiveMinimum', 'maximum', or 'exclusiveMaximum' supported."));
                return;
            }
            if (evaluationResult.value === true) {
                const resultObject = this.createResultObject(input_descriptor_idx, resultIdx, evaluationResult, results);
                results.push(resultObject);
            }
            else if (field.predicate === pex_models_1.Optionality.Required) {
                const resultObject = this.createWarnResultObject(input_descriptor_idx, resultIdx, evaluationResult, results, 'Predicate is required but not applied');
                results.push(resultObject);
            }
        }
    }
    retrieveResultInputDescriptorIdx(input_descriptor_path) {
        const inputDescriptorText = '$.input_descriptors[';
        const startIdx = input_descriptor_path.indexOf(inputDescriptorText);
        const startWithIdx = input_descriptor_path.substring(startIdx + inputDescriptorText.length);
        const endIdx = startWithIdx.indexOf(']');
        const idx = startWithIdx.substring(0, endIdx);
        return parseInt(idx);
    }
    concatenatePath(path) {
        let completePath = '';
        for (let i = 0; i < path.length; i++) {
            if (typeof path[i] === 'number') {
                completePath = completePath.substring(0, completePath.length - 1);
                completePath += '[*].';
            }
            else {
                completePath += path[i] + '.';
            }
        }
        return completePath.substring(0, completePath.length - 1);
    }
    createResultObject(input_descriptor_idx, resultIdx, evaluationResult, results) {
        return {
            input_descriptor_path: `$.input_descriptors[${input_descriptor_idx}]`,
            verifiable_credential_path: results[resultIdx].verifiable_credential_path,
            evaluator: this.getName(),
            status: ConstraintUtils_1.Status.INFO,
            message: Messages_1.default.INPUT_CANDIDATE_IS_ELIGIBLE_FOR_PRESENTATION_SUBMISSION,
            payload: evaluationResult,
        };
    }
    createWarnResultObject(input_descriptor_idx, resultIdx, evaluationResult, results, message) {
        return {
            input_descriptor_path: `$.input_descriptors[${input_descriptor_idx}]`,
            verifiable_credential_path: results[resultIdx].verifiable_credential_path,
            evaluator: this.getName(),
            status: ConstraintUtils_1.Status.WARN,
            message,
            payload: evaluationResult,
        };
    }
    createErrorResultObject(input_descriptor_idx, resultIdx, evaluationResult, results, message) {
        return {
            input_descriptor_path: `$.input_descriptors[${input_descriptor_idx}]`,
            verifiable_credential_path: results[resultIdx].verifiable_credential_path,
            evaluator: this.getName(),
            status: ConstraintUtils_1.Status.ERROR,
            message,
            payload: evaluationResult,
        };
    }
}
exports.PredicateRelatedFieldEvaluationHandler = PredicateRelatedFieldEvaluationHandler;
