// @credo-ts/didcomm/build/modules/credentials/models/DidCommCredentialState.mjs
//#region src/modules/credentials/models/DidCommCredentialState.ts
/**
* Issue Credential states as defined in RFC 0036 and RFC 0453
*
* @see https://github.com/hyperledger/aries-rfcs/blob/master/features/0036-issue-credential/README.md#states
*/
let DidCommCredentialState = /* @__PURE__ */ function(DidCommCredentialState) {
	DidCommCredentialState["ProposalSent"] = "proposal-sent";
	DidCommCredentialState["ProposalReceived"] = "proposal-received";
	DidCommCredentialState["OfferSent"] = "offer-sent";
	DidCommCredentialState["OfferReceived"] = "offer-received";
	DidCommCredentialState["Declined"] = "declined";
	DidCommCredentialState["RequestSent"] = "request-sent";
	DidCommCredentialState["RequestReceived"] = "request-received";
	DidCommCredentialState["CredentialIssued"] = "credential-issued";
	DidCommCredentialState["CredentialReceived"] = "credential-received";
	DidCommCredentialState["Done"] = "done";
	DidCommCredentialState["Abandoned"] = "abandoned";
	return DidCommCredentialState;
}({});

//#endregion
export { DidCommCredentialState };
