// @credo-ts/core/build/modules/dids/domain/verificationMethod/Multikey.mjs


import { CredoError } from "../../../../error/CredoError.mjs";
import "../../../../error/index.mjs";
import { PublicJwk } from "../../../kms/jwk/PublicJwk.mjs";
import "../../../kms/index.mjs";

//#region src/modules/dids/domain/verificationMethod/Multikey.ts
const VERIFICATION_METHOD_TYPE_MULTIKEY = "Multikey";
/**
* Get a Multikey verification method.
*/
function getMultikey({ did, publicJwk, verificationMethodId }) {
	if (!verificationMethodId) verificationMethodId = `${did}#${publicJwk.fingerprint}`;
	return {
		id: verificationMethodId,
		type: VERIFICATION_METHOD_TYPE_MULTIKEY,
		controller: did,
		publicKeyMultibase: publicJwk.fingerprint
	};
}
/**
* Check whether a verification method is a Multikey verification method.
*/
function isMultikey(verificationMethod) {
	return verificationMethod.type === VERIFICATION_METHOD_TYPE_MULTIKEY;
}
/**
* Get a public jwk from a Multikey verification method.
*/
function getPublicJwkFromMultikey(verificationMethod) {
	if (!verificationMethod.publicKeyMultibase) throw new CredoError(`Missing publicKeyMultibase on verification method with type ${VERIFICATION_METHOD_TYPE_MULTIKEY}`);
	return PublicJwk.fromFingerprint(verificationMethod.publicKeyMultibase);
}

//#endregion
export { VERIFICATION_METHOD_TYPE_MULTIKEY, getMultikey, getPublicJwkFromMultikey, isMultikey };
