// @credo-ts/core/build/modules/dids/domain/verificationMethod/VerificationMethod.mjs


import { __decorateMetadata } from "../../../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorateMetadata.mjs";
import { __decorate } from "../../../../_virtual/_@oxc-project_runtime@0.110.0/helpers/decorate.mjs";
import { IsOptional, IsString } from "class-validator";

//#region src/modules/dids/domain/verificationMethod/VerificationMethod.ts
var VerificationMethod = class {
	constructor(options) {
		if (options) {
			this.id = options.id;
			this.type = options.type;
			this.controller = options.controller;
			this.publicKeyBase58 = options.publicKeyBase58;
			this.publicKeyBase64 = options.publicKeyBase64;
			this.publicKeyJwk = options.publicKeyJwk;
			this.publicKeyHex = options.publicKeyHex;
			this.publicKeyMultibase = options.publicKeyMultibase;
			this.publicKeyPem = options.publicKeyPem;
			this.blockchainAccountId = options.blockchainAccountId;
			this.ethereumAddress = options.ethereumAddress;
		}
	}
};
__decorate([IsString(), __decorateMetadata("design:type", String)], VerificationMethod.prototype, "id", void 0);
__decorate([IsString(), __decorateMetadata("design:type", String)], VerificationMethod.prototype, "type", void 0);
__decorate([IsString(), __decorateMetadata("design:type", String)], VerificationMethod.prototype, "controller", void 0);
__decorate([
	IsOptional(),
	IsString(),
	__decorateMetadata("design:type", String)
], VerificationMethod.prototype, "publicKeyBase58", void 0);
__decorate([
	IsOptional(),
	IsString(),
	__decorateMetadata("design:type", String)
], VerificationMethod.prototype, "publicKeyBase64", void 0);
__decorate([
	IsOptional(),
	IsString(),
	__decorateMetadata("design:type", String)
], VerificationMethod.prototype, "publicKeyHex", void 0);
__decorate([
	IsOptional(),
	IsString(),
	__decorateMetadata("design:type", String)
], VerificationMethod.prototype, "publicKeyMultibase", void 0);
__decorate([
	IsOptional(),
	IsString(),
	__decorateMetadata("design:type", String)
], VerificationMethod.prototype, "publicKeyPem", void 0);
__decorate([
	IsOptional(),
	IsString(),
	__decorateMetadata("design:type", String)
], VerificationMethod.prototype, "blockchainAccountId", void 0);
__decorate([
	IsOptional(),
	IsString(),
	__decorateMetadata("design:type", String)
], VerificationMethod.prototype, "ethereumAddress", void 0);

//#endregion
export { VerificationMethod };
