// @credo-ts/didcomm/build/modules/discover-features/DidCommDiscoverFeaturesModule.mjs
import { DidCommFeatureRegistry } from "../../DidCommFeatureRegistry.mjs";
import { DidCommMessageHandlerRegistry } from "../../DidCommMessageHandlerRegistry.mjs";
import { DidCommProtocol } from "../../models/features/DidCommProtocol.mjs";
import "../../models/index.mjs";
import { DidCommDiscoverFeaturesModuleConfig } from "./DidCommDiscoverFeaturesModuleConfig.mjs";
import { DidCommDiscoverFeaturesV1Service } from "./protocol/v1/DidCommDiscoverFeaturesV1Service.mjs";
import "./protocol/v1/index.mjs";
import { DidCommDiscoverFeaturesV2Service } from "./protocol/v2/DidCommDiscoverFeaturesV2Service.mjs";
import "./protocol/v2/index.mjs";
import { DidCommDiscoverFeaturesApi } from "./DidCommDiscoverFeaturesApi.mjs";

//#region src/modules/discover-features/DidCommDiscoverFeaturesModule.ts
var DidCommDiscoverFeaturesModule = class {
	constructor(config) {
		this.api = DidCommDiscoverFeaturesApi;
		this.config = new DidCommDiscoverFeaturesModuleConfig(config);
	}
	/**
	* Registers the dependencies of the discover features module on the dependency manager.
	*/
	register(dependencyManager) {
		dependencyManager.registerInstance(DidCommDiscoverFeaturesModuleConfig, this.config);
		dependencyManager.registerSingleton(DidCommDiscoverFeaturesV1Service);
		dependencyManager.registerSingleton(DidCommDiscoverFeaturesV2Service);
	}
	async initialize(agentContext) {
		const featureRegistry = agentContext.resolve(DidCommFeatureRegistry);
		const messageHandlerRegistry = agentContext.resolve(DidCommMessageHandlerRegistry);
		const v1DiscoverFeatureService = agentContext.resolve(DidCommDiscoverFeaturesV1Service);
		const v2DiscoverFeatureService = agentContext.resolve(DidCommDiscoverFeaturesV2Service);
		v1DiscoverFeatureService.register(messageHandlerRegistry);
		v2DiscoverFeatureService.register(messageHandlerRegistry);
		featureRegistry.register(new DidCommProtocol({
			id: "https://didcomm.org/discover-features/1.0",
			roles: ["requester", "responder"]
		}), new DidCommProtocol({
			id: "https://didcomm.org/discover-features/2.0",
			roles: ["requester", "responder"]
		}));
	}
};

//#endregion
export { DidCommDiscoverFeaturesModule };
