// @credo-ts/core/build/modules/vc/sd-jwt-vc/W3cV2SdJwtVerifiableCredential.mjs


import { CredoError } from "../../../error/CredoError.mjs";
import "../../../error/index.mjs";
import { MessageValidator } from "../../../utils/MessageValidator.mjs";
import { JsonTransformer } from "../../../utils/JsonTransformer.mjs";
import "../../../utils/index.mjs";
import { ClaimFormat } from "../models/ClaimFormat.mjs";
import { W3cV2Credential } from "../models/credential/W3cV2Credential.mjs";
import { decodeSdJwt } from "./W3cV2SdJwt.mjs";

//#region src/modules/vc/sd-jwt-vc/W3cV2SdJwtVerifiableCredential.ts
/**
* Represents a Verifiable Credential encoded as a SD-JWT.
*
* @see https://www.w3.org/TR/vc-jose-cose/#securing-with-sd-jwt
*/
var W3cV2SdJwtVerifiableCredential = class W3cV2SdJwtVerifiableCredential {
	constructor(options) {
		this.sdJwt = options.sdJwt;
		this.resolvedCredential = JsonTransformer.fromJSON(options.sdJwt.prettyClaims, W3cV2Credential, { validate: false });
		this.validate();
	}
	static fromCompact(compact) {
		return new W3cV2SdJwtVerifiableCredential({ sdJwt: decodeSdJwt(compact, ClaimFormat.SdJwtW3cVc) });
	}
	/**
	* The encoded version of this credential.
	*/
	get encoded() {
		return this.sdJwt.compact;
	}
	/**
	* The {@link ClaimFormat} of the credential.
	*
	* For W3C VC SD-JWT credentials this is always `vc+sd-jwt`.
	*/
	get claimFormat() {
		return ClaimFormat.SdJwtW3cVc;
	}
	/**
	* Validates the SD-JWT and the resolved credential.
	*/
	validate() {
		MessageValidator.validateSync(this.resolvedCredential);
		const sdJwt = this.sdJwt;
		const header = sdJwt.header;
		const payload = sdJwt.prettyClaims;
		if ("typ" in header && header.typ !== "vc+sd-jwt") throw new CredoError(`The provided W3C VC SD-JWT does not have the correct 'typ' header.`);
		if ("cyt" in header && header.cyt !== "vc") throw new CredoError(`The provided W3C VC SD-JWT does not have the correct 'cyt' header.`);
		const iss = header.iss ?? payload.iss;
		if (iss) {
			if (this.resolvedCredential.issuerId !== iss) throw new CredoError(`The provided W3C VC SD-JWT has both 'iss' and 'issuer' claims, but they differ.`);
		}
		if (payload.jti) {
			if (this.resolvedCredential.id && this.resolvedCredential.id !== payload.jti) throw new CredoError(`The provided W3C VC JWT has both 'jti' and 'id' claims, but they differ.`);
		}
		if (payload.sub) {
			if (!this.resolvedCredential.credentialSubjectIds.includes(payload.sub)) throw new CredoError(`The provided W3C VC SD-JWT has a 'sub' claim, but it does not match any credentialSubject.`);
		}
	}
};

//#endregion
export { W3cV2SdJwtVerifiableCredential };
